{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ce2ca986-b0d9-43b1-8928-00dfbf050a84",
   "metadata": {},
   "source": [
    "# Submodule2: \n",
    "## Collect and prepare sequence data and analysis:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4ef0389-80cf-40e1-ac52-c1e75b28ce0e",
   "metadata": {},
   "source": [
    "### 2.1 Demonstrate Efficient Methods for Sourcing Pathogen Sequences and Preparing Data for Phylogenetic Analysis\n",
    "Efficient methods for sourcing pathogen sequences and preparing data include:\n",
    "* Public Databases: Utilize public repositories like GenBank, EMBL, and DDBJ for sourcing high-quality pathogen sequences. These databases provide comprehensive and well-annotated genetic data.\n",
    "* Sequence Retrieval Tools: Use tools like Entrez Direct and Biopython to automate the retrieval of sequences from public databases, ensuring efficiency and reducing the risk of manual errors.\n",
    "* Data Cleaning and Preprocessing: Implement tools such as Trimmomatic for trimming low-quality reads and removing adapters from raw sequence data. This step is crucial for ensuring the integrity of the sequences used in downstream analysis.\n",
    "* Sequence Alignment: Use alignment tools like MAFFT or ClustalW to align sequences, correcting for gaps and mismatches to maximize homology. Proper alignment is essential for accurate phylogenetic tree construction.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbe7bffb-431b-4518-8c64-7ca50ea1e996",
   "metadata": {},
   "source": [
    "### 2.2 Discuss the Importance of Cloud-Based Storage Solutions in Managing Metagenomic Sequence Data\n",
    "Cloud-based storage solutions are vital for managing metagenomic sequence data due to:\n",
    "- **Scalability:** Cloud storage offers scalable solutions to handle the vast amounts of data generated by metagenomic studies. It allows researchers to store large datasets without worrying about local storage limitations.\n",
    "- **Accessibility:** Cloud storage ensures that data can be accessed from anywhere, facilitating collaboration among researchers across different geographical locations.\n",
    "- **Cost-Effectiveness:** Cloud services often operate on a pay-as-you-go model, making it cost-effective as researchers only pay for the storage and computational resources they use.\n",
    "- **Data Security and Backup:** Cloud providers offer robust security measures and automatic backups, protecting valuable data from loss or unauthorized access.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c493fef6-c6b5-4058-8aea-1a094d8feb1f",
   "metadata": {},
   "source": [
    "### 2.3 Explain How the Incorporation of Publicly Available Datasets from Reputable Metagenomic Databases Enhances the Depth of Analysis\n",
    "Incorporating publicly available datasets from reputable metagenomic databases enhances analysis by:\n",
    "- **Increased Data Volume:** Access to a large volume of data increases the statistical power and robustness of analyses, enabling more accurate and comprehensive studies.\n",
    "- **Comparative Analysis:** Public datasets provide a wealth of comparative data, allowing researchers to identify patterns, variations, and evolutionary trends across different studies and datasets.\n",
    "- **Validation and Reproducibility:** Using standardized, publicly available data ensures that results can be validated and reproduced by other researchers, enhancing the credibility of the findings.\n",
    "- **Resource Sharing:** Public databases foster a collaborative environment where researchers can share resources, tools, and data, accelerating the pace of scientific discovery."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5a30c19-3759-4f49-9c04-b924c1082095",
   "metadata": {},
   "source": [
    "### KEGG Dataset:\n",
    "Dataset 1: KEGG for Phylogenetic Tree\n",
    "Downloading KEGG Dataset\n",
    "**KEGG (Kyoto Encyclopedia of Genes and Genomes)** provides a wealth of data for understanding high-level functions and utilities of biological systems. To download KEGG data:\n",
    "1. Access KEGG Dataset: KEGG provides a website for data retrieval.\n",
    "    - https://www.genome.jp/kegg/seq/\n",
    "2. Downloaded file from FASTA sequence files section on the website.\n",
    "    - More resources on KEGG: https://www.genome.jp/kegg/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72341617-aed7-4fb7-9b7e-55cd78832952",
   "metadata": {},
   "source": [
    "## Download KEGG dataset from URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9ec32755-9a73-48bb-afbb-77cd40cb56f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -o \"./data/kegg/br01553.fasta\" \"https://firebasestorage.googleapis.com/v0/b/reactfirebase-142f5.appspot.com/o/nosi-phylogeny%2Fbr01553.fasta?alt=media&token=08d87736-6f5b-4512-94c8-16531c87aa38\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71b5f3e7-c8c0-4200-98c8-90333459d7a5",
   "metadata": {},
   "source": [
    "## Uniport Dataset\n",
    "Dataset 2: UniProt for Phylogenetic Tree\n",
    "Downloading UniProt Dataset\n",
    "UniProt is a comprehensive resource for protein sequence and functional information. To download UniProt data:\n",
    "1. Access UniProt Website: Visit the UniProt website and search for the desired protein sequences.\n",
    "2. Retrieve Data: Downloaded the Isoform sequences fasta file from the website.\n",
    "    - File name: uniprot_sprot_varsplic.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "561ef7c5-5ac9-4b39-9937-05fe514a2ffc",
   "metadata": {},
   "source": [
    "## Download Uniport Dataset from URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c405ffa8-cbfb-416a-afbb-4d12781e01cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -o \"./data/uniport/uniprot_sprot_varsplic.fasta\" \"https://firebasestorage.googleapis.com/v0/b/reactfirebase-142f5.appspot.com/o/uniprot_sprot_varsplic.fasta?alt=media&token=a7695fee-82a3-4100-983c-5eae5db474e9\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbd3190a-88b9-4d99-8b84-ea7f6c45e2bf",
   "metadata": {},
   "source": [
    "### SARS-CoV-2 Dataset\n",
    "Downloaded SARS-CoV-2 sequence data from the Nextstrain project: \n",
    "1. https://docs.nextstrain.org/projects/nextclade/en/stable/user/datasets.html\n",
    "2. The specific file used is: data/nextstrain/sars-cov-2/wuhan-hu-1/proteins/sequences.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "550bbceb-0775-45be-9f13-7a894e40965b",
   "metadata": {},
   "source": [
    "## Download SARS-Cov-2 Dataset from URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fa87886-458d-433b-8f24-f715f0995fa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -o \"./data/cov/sequences.fasta\" \"https://firebasestorage.googleapis.com/v0/b/reactfirebase-142f5.appspot.com/o/nosi-phylogeny%2Fsequences.fasta?alt=media&token=4de7706b-0084-4846-b829-79d8ea35d327\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b0fb3b4-9e81-4228-a159-a5bf2b1a9179",
   "metadata": {},
   "source": [
    "### Data Preparation\n",
    "**Convert the .fasta file to .fastq format using the following Python code:**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e74d5fb-729c-4494-b2ee-7330a8be297b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "def fasta_to_fastq(fasta_file, fastq_file, quality=40):\n",
    "    with open(fastq_file, \"w\") as output_handle:\n",
    "        for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "            record.letter_annotations[\"phred_quality\"] = [quality] * len(record.seq)\n",
    "            SeqIO.write(record, output_handle, \"fastq\")\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    fasta_file = \"data/cov/sequences.fasta\"\n",
    "    fastq_file = \"data/cov/sequences_converted.fastq\"\n",
    "    fasta_to_fastq(fasta_file=fasta_file, fastq_file=fastq_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c050cec7-becd-44b4-bc3b-9c94ef8307a6",
   "metadata": {},
   "source": [
    "### 2.4 Implement Quality Control Checks Using Tools Like MultiQC and FastQC to Ensure Data Integrity\n",
    "**Using FASTQC for Quality Control.**\n",
    "- Follow installation and setup from:\n",
    "    - https://raw.githubusercontent.com/s-andrews/FastQC/master/INSTALL.txt\n",
    "- First need to install JAVA as FastQC uses JAVA.\n",
    "- RUN FASTQC command line:\n",
    "    - \tfastqc -t 4./data/uniport/ uniprot_sprot_varsplic.fastq\r\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51249d58-662f-447c-9c99-e7cd74557184",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
