{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ce2ca986-b0d9-43b1-8928-00dfbf050a84",
   "metadata": {},
   "source": [
    "# Submodule2: \n",
    "## Collect and prepare sequence data and analysis:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4ef0389-80cf-40e1-ac52-c1e75b28ce0e",
   "metadata": {},
   "source": [
    "### 2.1 Demonstrate Efficient Methods for Sourcing Pathogen Sequences and Preparing Data for Phylogenetic Analysis\n",
    "Efficient methods for sourcing pathogen sequences and preparing data include:\n",
    "* Public Databases: Utilize public repositories like GenBank, EMBL, and DDBJ for sourcing high-quality pathogen sequences. These databases provide comprehensive and well-annotated genetic data.\n",
    "* Sequence Retrieval Tools: Use tools like Entrez Direct and Biopython to automate the retrieval of sequences from public databases, ensuring efficiency and reducing the risk of manual errors.\n",
    "* Data Cleaning and Preprocessing: Implement tools such as Trimmomatic for trimming low-quality reads and removing adapters from raw sequence data. This step is crucial for ensuring the integrity of the sequences used in downstream analysis.\n",
    "* Sequence Alignment: Use alignment tools like MAFFT or ClustalW to align sequences, correcting for gaps and mismatches to maximize homology. Proper alignment is essential for accurate phylogenetic tree construction.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbe7bffb-431b-4518-8c64-7ca50ea1e996",
   "metadata": {},
   "source": [
    "### 2.2 Discuss the Importance of Cloud-Based Storage Solutions in Managing Metagenomic Sequence Data\n",
    "Cloud-based storage solutions are vital for managing metagenomic sequence data due to:\n",
    "- **Scalability:** Cloud storage offers scalable solutions to handle the vast amounts of data generated by metagenomic studies. It allows researchers to store large datasets without worrying about local storage limitations.\n",
    "- **Accessibility:** Cloud storage ensures that data can be accessed from anywhere, facilitating collaboration among researchers across different geographical locations.\n",
    "- **Cost-Effectiveness:** Cloud services often operate on a pay-as-you-go model, making it cost-effective as researchers only pay for the storage and computational resources they use.\n",
    "- **Data Security and Backup:** Cloud providers offer robust security measures and automatic backups, protecting valuable data from loss or unauthorized access.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c493fef6-c6b5-4058-8aea-1a094d8feb1f",
   "metadata": {},
   "source": [
    "### 2.3 Explain How the Incorporation of Publicly Available Datasets from Reputable Metagenomic Databases Enhances the Depth of Analysis\n",
    "Incorporating publicly available datasets from reputable metagenomic databases enhances analysis by:\n",
    "- **Increased Data Volume:** Access to a large volume of data increases the statistical power and robustness of analyses, enabling more accurate and comprehensive studies.\n",
    "- **Comparative Analysis:** Public datasets provide a wealth of comparative data, allowing researchers to identify patterns, variations, and evolutionary trends across different studies and datasets.\n",
    "- **Validation and Reproducibility:** Using standardized, publicly available data ensures that results can be validated and reproduced by other researchers, enhancing the credibility of the findings.\n",
    "- **Resource Sharing:** Public databases foster a collaborative environment where researchers can share resources, tools, and data, accelerating the pace of scientific discovery."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f104290f-4fc9-4e61-9473-968c03be0f3d",
   "metadata": {},
   "source": [
    "# Create data directory to store data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8fa8e9f8-1279-4bbf-b4a4-6535e34f5737",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "data_dir = os.path.isdir('./data')\n",
    "print(data_dir)\n",
    "if not data_dir:\n",
    "    !mkdir data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5a30c19-3759-4f49-9c04-b924c1082095",
   "metadata": {},
   "source": [
    "### KEGG Dataset:\n",
    "Dataset 1: KEGG for Phylogenetic Tree\n",
    "Downloading KEGG Dataset\n",
    "**KEGG (Kyoto Encyclopedia of Genes and Genomes)** provides a wealth of data for understanding high-level functions and utilities of biological systems. To download KEGG data:\n",
    "1. Access KEGG Dataset: KEGG provides a website for data retrieval.\n",
    "    - https://www.genome.jp/kegg/seq/\n",
    "2. Downloaded file from FASTA sequence files section on the website.\n",
    "    - More resources on KEGG: https://www.genome.jp/kegg/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72341617-aed7-4fb7-9b7e-55cd78832952",
   "metadata": {},
   "source": [
    "## Download KEGG dataset from URL"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2b670e3-7ae3-4664-9c66-767993bb0155",
   "metadata": {},
   "source": [
    "### Create Kegg folder to store data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bea11637-3a5f-4ce6-a874-f38f681d2ae0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Check if the directory exists\n",
    "kegg_dir = os.path.isdir('./data/kegg')\n",
    "print(kegg_dir)\n",
    "\n",
    "# If the directory does not exist, create it\n",
    "if not kegg_dir:\n",
    "    try:\n",
    "        os.makedirs('./data/kegg/')\n",
    "        print(\"Directory created successfully\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9ec32755-9a73-48bb-afbb-77cd40cb56f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -o \"./data/kegg/br01553.fasta\" \"https://firebasestorage.googleapis.com/v0/b/reactfirebase-142f5.appspot.com/o/nosi-phylogeny%2Fbr01553.fasta?alt=media&token=08d87736-6f5b-4512-94c8-16531c87aa38\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71b5f3e7-c8c0-4200-98c8-90333459d7a5",
   "metadata": {},
   "source": [
    "## Uniport Dataset\n",
    "Dataset 2: UniProt for Phylogenetic Tree\n",
    "Downloading UniProt Dataset\n",
    "UniProt is a comprehensive resource for protein sequence and functional information. To download UniProt data:\n",
    "1. Access UniProt Website: Visit the UniProt website and search for the desired protein sequences.\n",
    "2. Retrieve Data: Downloaded the Isoform sequences fasta file from the website.\n",
    "    - File name: uniprot_sprot_varsplic.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "561ef7c5-5ac9-4b39-9937-05fe514a2ffc",
   "metadata": {},
   "source": [
    "## Download Uniport Dataset from URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "22071134-8b6f-4dcd-8d34-2703cc064699",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "Directory created successfully\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Check if the directory exists\n",
    "uniport_dir = os.path.isdir('./data/uniport')\n",
    "print(uniport_dir)\n",
    "\n",
    "# If the directory does not exist, create it\n",
    "if not uniport_dir:\n",
    "    try:\n",
    "        os.makedirs('./data/uniport/')\n",
    "        print(\"Directory created successfully\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c405ffa8-cbfb-416a-afbb-4d12781e01cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -o \"./data/uniport/uniprot_sprot_varsplic.fasta\" \"https://firebasestorage.googleapis.com/v0/b/reactfirebase-142f5.appspot.com/o/uniprot_sprot_varsplic.fasta?alt=media&token=a7695fee-82a3-4100-983c-5eae5db474e9\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbd3190a-88b9-4d99-8b84-ea7f6c45e2bf",
   "metadata": {},
   "source": [
    "### SARS-CoV-2 Dataset\n",
    "Downloaded SARS-CoV-2 sequence data from the Nextstrain project: \n",
    "1. https://docs.nextstrain.org/projects/nextclade/en/stable/user/datasets.html\n",
    "2. The specific file used is: data/nextstrain/sars-cov-2/wuhan-hu-1/proteins/sequences.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "550bbceb-0775-45be-9f13-7a894e40965b",
   "metadata": {},
   "source": [
    "## Download SARS-Cov-2 Dataset from URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "150d58da-941e-4112-8020-4a3fcbc5e6a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "Directory created successfully\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Check if the directory exists\n",
    "cov_dir = os.path.isdir('./data/cov')\n",
    "print(cov_dir)\n",
    "\n",
    "# If the directory does not exist, create it\n",
    "if not cov_dir:\n",
    "    try:\n",
    "        os.makedirs('./data/cov/')\n",
    "        print(\"Directory created successfully\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fa87886-458d-433b-8f24-f715f0995fa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -o \"./data/cov/sequences.fasta\" \"https://firebasestorage.googleapis.com/v0/b/reactfirebase-142f5.appspot.com/o/nosi-phylogeny%2Fsequences.fasta?alt=media&token=4de7706b-0084-4846-b829-79d8ea35d327\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b0fb3b4-9e81-4228-a159-a5bf2b1a9179",
   "metadata": {},
   "source": [
    "### Data Preparation\n",
    "**Convert the .fasta file to .fastq format using the following Python code:**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3673dc77-b348-4f50-992d-4d4dbc4405ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting biopython\n",
      "  Obtaining dependency information for biopython from https://files.pythonhosted.org/packages/89/38/e45df36e10de29141ba0e5d94bfb942925e62cb6ad182e58b74fa5edcfcc/biopython-1.84-cp311-cp311-win_amd64.whl.metadata\n",
      "  Downloading biopython-1.84-cp311-cp311-win_amd64.whl.metadata (13 kB)\n",
      "Requirement already satisfied: numpy in c:\\users\\anush\\anaconda3\\lib\\site-packages (from biopython) (1.26.4)\n",
      "Downloading biopython-1.84-cp311-cp311-win_amd64.whl (2.8 MB)\n",
      "   ---------------------------------------- 0.0/2.8 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.0/2.8 MB 330.3 kB/s eta 0:00:09\n",
      "    --------------------------------------- 0.1/2.8 MB 656.4 kB/s eta 0:00:05\n",
      "   -- ------------------------------------- 0.2/2.8 MB 1.5 MB/s eta 0:00:02\n",
      "   -------- ------------------------------- 0.6/2.8 MB 3.6 MB/s eta 0:00:01\n",
      "   ----------------------- ---------------- 1.6/2.8 MB 7.5 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 2.7/2.8 MB 10.8 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.8/2.8 MB 9.9 MB/s eta 0:00:00\n",
      "Installing collected packages: biopython\n",
      "Successfully installed biopython-1.84\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e74d5fb-729c-4494-b2ee-7330a8be297b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "def fasta_to_fastq(fasta_file, fastq_file, quality=40):\n",
    "    with open(fastq_file, \"w\") as output_handle:\n",
    "        for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "            record.letter_annotations[\"phred_quality\"] = [quality] * len(record.seq)\n",
    "            SeqIO.write(record, output_handle, \"fastq\")\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    fasta_file = \"./data/cov/sequences.fasta\"\n",
    "    fastq_file = \"./data/cov/sequences_converted.fastq\"\n",
    "    fasta_to_fastq(fasta_file=fasta_file, fastq_file=fastq_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2a10fb4-2d75-40b5-8ceb-9fc6b116aa5a",
   "metadata": {},
   "source": [
    "### 2.4 Implement Quality Control Checks Using Tools Like MultiQC and FastQC to Ensure Data Integrity\n",
    "**Using FASTQC for Quality Control.**\n",
    "- FastQC is a java application.  In order to run it needs your system to have a suitable Java Runtime Environment (JRE) installed.  Before you try to run FastQC you should therefore ensure that you have a suitable JRE.  There are a number of different JREs available however the ones we have tested are the latest Oracle runtime environments and those from the adoptOpenJDK project (https://adoptopenjdk.net/).  You need to download and install a suitable 64-bit JRE and make sure that the java application is in your path (most installers will take care of this for you).\n",
    "\n",
    "On linux most distributions will have java installed already so you might not need to do anything.  If java isn't installed then you can add it by doing:\n",
    "\n",
    "- Ubuntu / Mint: **sudo apt install default-jre**\n",
    "\n",
    "- CentOS / Redhat: **sudo yum install java-1.8.0-openjdk**\n",
    "\n",
    "You can check whether java is installed by opening the 'cmd' program on windows, or any shell on linux and typing:\n",
    "\n",
    "**java --version**\n",
    "\n",
    "You should see something like:\n",
    "\n",
    "- java -version\n",
    "- openjdk version \"11.0.2\" 2019-01-15\n",
    "- OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.2+9)\n",
    "- OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.2+9, mixed mode)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8ac211c-d14e-4aea-98a5-f1c68cae643a",
   "metadata": {},
   "source": [
    "### Installation of FastQC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5e65e01-d60c-42d0-8f2d-33f34dde4fd5",
   "metadata": {},
   "source": [
    "### Download FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "51249d58-662f-447c-9c99-e7cd74557184",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10445846-8798-4740-bb95-62aefa58c0cc",
   "metadata": {},
   "source": [
    "### Extract the Downloaded File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c3a5801-2635-45bb-b75a-477b9fda15a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "!unzip fastqc_v0.11.9.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6b4453f-5ab0-42ba-a791-22595baf3c85",
   "metadata": {},
   "source": [
    "### Make the FastQC Executable\n",
    "Navigate to the FastQC directory and make the fastqc file executable:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76565fbb-f59b-47d0-903c-4367675a940d",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cd FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcca8708-e663-4706-bfac-79eed92a3b31",
   "metadata": {},
   "outputs": [],
   "source": [
    "!chmod +x fastqc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd4c220f-851a-4f6b-8bfe-b6fc8cf7a70e",
   "metadata": {},
   "source": [
    "### Verify the Installation\n",
    "To verify that FastQC is installed correctly, you can check its version:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "970e4d58-1d66-4308-81d0-a2e213e23f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "!fastqc --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d13e9c8-6e42-4af3-9656-69023e80b9bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Using FastQC\n",
    "You can now use FastQC to analyze your sequence data. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dea022ee-cab1-4d8b-82ae-2b1682551362",
   "metadata": {},
   "outputs": [],
   "source": [
    "!fastqc -t 4 uniprot_sprot_varsplic.fastq"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
