{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ead4567e-3195-474f-b344-bc36cb46ac4c",
   "metadata": {},
   "source": [
    "# Submodule3: Construct Phylogenetic Tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b36c7550",
   "metadata": {},
   "source": [
    "# Learning Objectives:\n",
    "In submodule 3 we will construct phylogenetic tree from a gene sequence that includes the following steps:\n",
    "- Perform sequence alignment\n",
    "- Perform phylogenetic tree reconstruction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f9ad573",
   "metadata": {},
   "source": [
    "----------------------------------------------------------------------------------------------------------------\n",
    "# Training Plan \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "097d5864",
   "metadata": {},
   "source": [
    "Submodule #1: Understanding the Basics of Phylogenetic\n",
    "\n",
    "Submodule #2: Collect and Prepare Sequence Data and Analysis\n",
    "\n",
    "<font color=\"green\"> **Submodule #3: Construct Phylogenetic Tree** </font>\n",
    " \n",
    "Submodule #4: Analyze Phylogenetic Tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7cad5e6",
   "metadata": {},
   "source": [
    "----------------------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "attachments": {
    "17eaadba-d4ac-409f-be74-9ad593702af2.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAAK8AAAAQCAYAAABgDi/VAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAXfSURBVGhD7ZrPS1xJEMe/u//DoANRMMzF4yhkCLkYWUECMh4kewlIxERBInvb2eSQuSjjJZAhEiWyEOJFEeIgcYVA4mElGKIDe3EPQw5jyGj8I95Wd1e/1/1+zZiYrML7wGD369dd1dXdVdUPf3IIJCRcQH7mvwkJF44ftnn3yxO4XK4GyvFUMTswgdk9rvo4WS/i8vQmTrh+GmRfGlv8osb/euL1Pnu+o7zjTYwNFLF6zHWDb7H/WfDDNm/P9CI+Tme59n/TwNs3DYyXSKfXi7jfy48TLhShm1d6RvZK8hfjMeXp08/2Fux+AwvYVy0x3raB1WmzzwTG1hvcxuwY48addJ/8cE8kvFQRhQNgqWCOp7yX7mvqIHSfXRceSLTFeCHdv7zLTzW+Obp2CHrMgJ2E5yMd/5V23rTGCdjJJUqeImp9NWb72Eqdn0ZRx7NInXx6mGvH81rVslgHy44RtnYRFzaTvcd3na57r5wvXHc+PHW6frnrzHzQ5afOnmohPjsr98y2h87KkWpxnH1nhvp1Pd6XNTluoKz66+eCLy8fGuPwGIZMUz/5rtbVr9vRK+e2pY+JobdEybn98nNoXcq05m2jdPbpqG0WmKNdF+/65Zr2F2OLdiXD0Nmyt+rXqrzI9SXsdu4bYcd4nVRfb248ttZLro/dbq2nwL+mPgKeV4b38g2kuI7eHMa5qMpVbGlPcUzlgywGRdjtnaQQXMTNNtUE0PMhLkaSxs2ynU6krl5BH5c146VJ9HC559c8+g7e4611IumEP69a76HtBuamgMKK7VXCOFmvYKk7j7nhND/J4n4pi+35ihs5MJTzxrZQKUjfVN7TcXrSs9lehbx8FmvuHGnOD2gOGxXpVXqukZw3VeWRjhuodWepr56fGBsYvMp6kY53dIoj14Le/8R1TTN5cesr1naD7D2q27mvLEcQpZPQA6ZN2S6shyLtzY1kP5sHSg9M3SaxNlTFk4gIE5nzeu57AUv8TG/IpR128e/eY9u/qDLBV25/hAzRKm6oulXBNj9TpJG5xEVBG9UDi3aEmk4DWLb45eajwqrNYZ3ey6Q9owkudcYvmouSnenwFgloR6ZblU4Oj2iBO9Ghqgo5B0bIOajjkIrCnujPYzDD8xPOAVdw3XUIzWkqjwldX3F4Qu19eqQeBxXkjPWw95IPKbuBwi3z/fg9FNi8elK5+Xas0WXm42vDixDCU2BjlzwSX3pEXaA3LW2+DF+E1pp6Xm/TjtTy2BXylpuc9Bj0Bcz6nZtLYgRt5BC6VTQTh0gcAmFj6SA+Ub7Zn7UP1TfSbH3PFPLKck2tnxmd/aRRWva/v4g/De9t4tu85i3cCMEmOnVYN1IGQnlhkTqc5vYuwhQrbIYyC5+XDfMO7Olqh615Wj8dnWScWsO+CNLGsSNAFGGylTcWpDrayQMpz+oi56BJ43p/mvpvki3YnsIbkz6rO0dGWG2NeHlN1jcsqlm6tk6oHnGERtR4QtMGbyHEbdHv6jl1oHxw2x+ejA0gTnhraYOpMN2+A2kDyXqub6mkzwy1D+V9p5dys1GRoy4a+ZS66UbfyD1Sw3mMU4j73X2X9ChUrTw2Gi3by4/3y4bNevMokWcdcW/0wTmIhRb9l7Q9xUKSPoXa6VIGSQvyotc3iztT6aC9ZfmUBPQg5Ncg7wuUDcsumO3BrzEmvs2rEnTM63yoiNpokZTw8lyBTB2IPiOkpYYnUIKX4+TqFDJIGZViRKEuRl6uSinHsghjDWy90wYmz9xf53GL8hKwG5YKUHK/Ky5obs5UxFZ/MTLk2JAe5Iky7rwXUJtqtS8hZR9hRPadwJNOOgzcJG1aJhvW9Gc8mkOGIpQ5B740efZUDiKQh7dEnLzm65saLmIto9eR2ikH/7o0TughLmjG58sCKFWJiOiElE0XNG1HsQ6gKKEi+T94lM3ikXEWkv9tSLiwWJv3j0cvsPLX31xLSDgfXOvpxou537jmkXjehAtL5HfehITzDfAfi3Y7Mwn4qFgAAAAASUVORK5CYII="
    },
    "e69eac74-340a-4a4c-b1c6-d884188a09e4.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAALEAAAAaCAYAAAD48r3oAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAWBSURBVHhe7Zq/TxtJFMffBnMhSOSSzieRwtL9BQcFREqDxB8QI0cCqjRQpKOFu8JAd6K7AqTTVRgpFtwfgESDZLsAmrSRXATJriAYBTjiZO+92Te7s2vverxeTJDmI412Z3Z+7ezbt983tmUjYDA8YB7x0WB4sBgjNjx4IuXE+Pg4n+lzdHTEZwZDf+hoxLpGKet208ZgSILE5YQ0ZIOhX3RlxFtbWyJZltWSVIwhG/qJthGT8S4uLoqkgzFkQ79ITE4cHx8LoyWvLI9E0Et3RaWB7RtQ5mz/uIJ16xzWK5xNitNLmMF+LZE+Q+GUyyNx5mKtXnE+Wcqrd9d318R93hTYhTE2NsZnIvjzpc3NTVFOx+C1YIpF+QLbnmG6sEtc1B++2Gti3DN7rcxFiYD95hp2jXOlPI1xbm9/4oJ2fGrYWZ4L5L9woSGI9u5E0KOqzYLX0LD5DIT8iBgiGnozJwFK9lPAQ2+gF1w/HITl2SEuiIK833/opp7D8gQX9Uh95xKqsyPKfThjnBSewG6HOZG3fAmPwf59mEsMKtpyggxRTRLSyj8+N1BYasIJ5+6DtM+AiUeQyfGpoSdia2K5S9Eu0FtYWBBJDzSwN1InRugzoZeojtSSrBUxedpVllEdPKe+hA69hvkiwN7ctdfep08xvbmEutNJAG8cR6/55+uM7ZXN7NyIVp35DtWiBTOvdL4MOqjzcu5R6F2ep5yXWybW2WnjzZn7oGvuekfFBhH1lfX1r0mbZ8TUdz5HPIcI0KuGompilSgdLLUyIeuFc21v587sbOHaybIOdrWoyCuaWORVHenoV1mfdKZz7vTr6UinnjtOcFzWnp4G9vfbUh+pFc5tUDQu9bHt1teA7kVT5wr9HLducA3x3rbFffAauevi5VUNLvpT79Mloj7lZZvA+GHPSKyn2q4LYnnisG020sL6Hhip3MI8pOAvqQknnqLF96ZDVzboTR6CuffPIzSkc11oUfIgL5qwx1faMwRTry3Y+/er30sUv8EB7zDUDwEy2vNGD7aBWv8ONO7k9ADAH81AhI/zlN608g1AeH9agyew7UoaJZ/39Hfm17DdpYj6uRTU3o9AWpQQ+NVRdmLaPaP07DOoFeLtZCWyxUbGiy9EdwaM1Kvf+SwZJt+mIFtswi/u54ovtEN+7vZTOPfHsMbFYaRfDWDf0miv4O+PP0Epb8P8P85n+QAGWPOq8oNS65ZReRU/rxu9Bavi06uM436yJ1J4L9Jo8WXZt6CExuEYDo5dBZgapWv3Q1fPSJNEjLhb45WkMzh80QZc12QYHYFd9OSOUaKBvQhbJDQ09L6/laO8dYDRQZjJ2bB7iMaC3izzdtjzeqdf0QXJfoZhWcxBJr+xkvEdTD+DuR4NiTyXNwZ/VQTDMJVHb7ePLxfNaxoDSvcFvMK1HlQ85D2g/Yz0iWXE5HXVFBv2Gi8VcS+2wnyBgPIpGrUgizcu8/WdW1jB48qkE0yUV+WCkCGp3lXZCahcuot2UuVxKk3RTxD3ukBKiltYrw44Rsjz/3NJT0qQAb/DflS5VF7ttLl/A9UPfKqJfLnWUeJM0VjyBVxqspTws/cx2S9iFOHPiInj1NAIQwkL7KiZmsLoHNgR3o8LTuIgQAQEsswL5tzAAVO20AgEdud2Vgk21EDMbRcMQETZhTsH2ae8pvbRGgByP2ogFYJvPDW5bXkd1L7UHztE0v3hJxjItglEWwIzDEwDgZobbIlEY3Mb0U9rYNdSv80zDHtGrWPpE+uvmFE/fKjI/1tEDGEw9My9GjEFJAZDO2q1nyGd1lO7sYxYF+OJDf0gkd0Jg+E+6eiJibhbaOSJ6S+axhMb7pJII5YENXC3GCM23CVaRmww/LgA/A9Gb1+22FfYCwAAAABJRU5ErkJggg=="
    }
   },
   "cell_type": "markdown",
   "id": "939082d7-f377-4e20-9837-8da69673f171",
   "metadata": {},
   "source": [
    "## 3.1 Perform Accurate Sequence Alignment of Metagenomic Data using ClustalW\n",
    "Sequence alignment is a critical step in phylogenetic analysis, as it arranges the sequences in a manner that highlights their similarities and differences, allowing for accurate tree construction.\n",
    "Using ClustalW for Sequence Alignment:\n",
    "1. Download Clustal\n",
    "    - Obtain the ClustalW tool from its official website: http://www.clustal.org/clustal2/#Download\n",
    "    - click on ![image.png](attachment:17eaadba-d4ac-409f-be74-9ad593702af2.png) link in the webpage.\n",
    "    - This will take you to another page where you need to download for Windows ![image.png](attachment:e69eac74-340a-4a4c-b1c6-d884188a09e4.png)\n",
    "    - Once downloaded double-click on the downloaded file and complete the installation process.\n",
    "2. Install Clustal\n",
    "    - Follow the installation instructions specific to your operating system. For example, on Windows, it is typically installed at:\n",
    "        - C:\\Program Files (x86)\\ClustalW2\\clustalw2.exe\n",
    "3. Run ClustalW using Python and Biopyon:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2563a93f-ae2d-4b65-8f49-c88fe4a3fe11",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: matplotlib in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (3.8.4)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib) (1.2.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib) (4.53.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib) (1.4.5)\n",
      "Requirement already satisfied: numpy>=1.21 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib) (1.22.4)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib) (21.3)\n",
      "Requirement already satisfied: pillow>=8 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib) (10.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib) (3.1.2)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib) (2.9.0)\n",
      "Requirement already satisfied: six>=1.5 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib) (1.16.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install matplotlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "709e43dd-9921-4293-864a-82e7b86388e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: networkx in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (3.3)\r\n"
     ]
    }
   ],
   "source": [
    "!pip install networkx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "736ed8e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (1.84)\n",
      "Requirement already satisfied: numpy in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from biopython) (1.22.4)\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "098c393a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/ec2-user/SageMaker/nosi-phylogenetic-tree'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "74479142-d135-48cc-bfaf-ce166caf3640",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/SageMaker/nosi-phylogenetic-tree/data/cov\n"
     ]
    }
   ],
   "source": [
    "%cd data/cov/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "deb469d8-ba44-4189-9a38-c19b0c50d39e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/ec2-user/SageMaker/nosi-phylogenetic-tree/data/cov'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fceeb13",
   "metadata": {},
   "source": [
    "## Install and locate clustalw for sequence alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2770e362",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 'conda-forge' already in 'channels' list, moving to the top\n",
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: - \n",
      "The environment is inconsistent, please check the package plan carefully\n",
      "The following packages are causing the inconsistency:\n",
      "\n",
      "  - conda-forge/noarch::autopep8==2.0.4=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::black==24.4.2=py310hff52083_0\n",
      "  - conda-forge/noarch::bleach==6.1.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::plotly==5.23.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::pytest==8.3.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::qtpy==2.4.1=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::sip==6.7.12=py310hc6cd4ac_0\n",
      "  - conda-forge/noarch::flask==3.0.3=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::importlib_metadata==8.2.0=hd8ed1ab_0\n",
      "  - conda-forge/noarch::lazy_loader==0.4=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::pyqt5-sip==12.12.2=py310hc6cd4ac_5\n",
      "  - conda-forge/noarch::pytoolconfig==1.2.5=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::qdarkstyle==3.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::qtawesome==1.3.1=pyh9208f05_0\n",
      "  - conda-forge/noarch::yapf==0.40.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::dask-core==2024.7.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::flask-cors==4.0.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyter_client==8.6.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::rope==1.13.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::distributed==2024.7.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::ipykernel==6.29.5=pyh3099207_0\n",
      "  - conda-forge/noarch::keyring==25.2.1=pyha804496_0\n",
      "  - conda-forge/noarch::python-lsp-server==1.7.4=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyter_console==6.6.3=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::nbclient==0.10.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::pyls-spyder==0.4.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::python-lsp-black==2.0.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::qtconsole-base==5.4.4=pyha770c72_0\n",
      "  - conda-forge/noarch::s3transfer==0.10.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::spyder-kernels==2.4.4=unix_pyh707e725_0\n",
      "  - conda-forge/noarch::nbconvert-core==7.16.4=pyhd8ed1ab_1\n",
      "  - conda-forge/linux-64::pyqt==5.15.9=py310h04931ad_5\n",
      "  - conda-forge/linux-64::astropy==6.1.0=py310h8a78493_0\n",
      "  - conda-forge/noarch::bokeh==3.5.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyter_server==2.14.2=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::matplotlib-base==3.8.4=py310hef631a5_2\n",
      "  - conda-forge/noarch::nbconvert-pandoc==7.16.4=hd8ed1ab_1\n",
      "  - conda-forge/linux-64::pyqtwebengine==5.15.9=py310h704022c_5\n",
      "  - conda-forge/linux-64::pytables==3.9.2=py310h73b55d5_3\n",
      "  - conda-forge/noarch::qtconsole==5.4.4=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::statsmodels==0.14.2=py310h261611a_0\n",
      "  - conda-forge/noarch::jupyter-lsp==2.2.5=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyterlab_server==2.27.3=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::matplotlib==3.8.4=py310hff52083_2\n",
      "  - conda-forge/noarch::nbconvert==7.16.4=hd8ed1ab_1\n",
      "  - conda-forge/noarch::notebook-shim==0.2.4=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::scikit-image==0.24.0=py310hf9f9076_1\n",
      "  - conda-forge/noarch::seaborn-base==0.13.2=pyhd8ed1ab_2\n",
      "  - conda-forge/linux-64::shap==0.45.1=cpu_py310hdd52748_0\n",
      "  - conda-forge/noarch::jupyterlab==4.2.4=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::seaborn==0.13.2=hd8ed1ab_2\n",
      "  - conda-forge/noarch::dask-expr==1.1.9=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::notebook==7.2.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::dask==2024.7.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyter==1.0.0=pyhd8ed1ab_10\n",
      "  - conda-forge/noarch::hdijupyterutils==0.21.0=pyh1a96a4e_0\n",
      "  - conda-forge/noarch::autovizwidget==0.21.0=pyh1a96a4e_0\n",
      "  - conda-forge/noarch::sparkmagic==0.21.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::numpydoc==1.7.0=pyhd8ed1ab_3\n",
      "  - conda-forge/linux-64::spyder==5.4.5=py310hff52083_0\n",
      "  - conda-forge/noarch::sphinxcontrib-applehelp==2.0.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinxcontrib-devhelp==2.0.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinxcontrib-htmlhelp==2.1.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinxcontrib-qthelp==2.0.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinx==8.0.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinxcontrib-serializinghtml==1.1.10=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinxcontrib-websupport==1.2.7=pyhd8ed1ab_0\n",
      "done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 23.3.1\n",
      "  latest version: 24.7.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=24.7.1\n",
      "\n",
      "\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/ec2-user/anaconda3/envs/python3\n",
      "\n",
      "  added / updated specs:\n",
      "    - clustalw\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    clustalw-2.1               |      h4ac6f70_10         339 KB  bioconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:         339 KB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  botocore           conda-forge/noarch::botocore-1.35.23-pyge310_1234567_0 \n",
      "  cloudpickle        conda-forge/noarch::cloudpickle-3.0.0-pyhd8ed1ab_0 \n",
      "  clustalw           bioconda/linux-64::clustalw-2.1-h4ac6f70_10 \n",
      "  docutils           conda-forge/noarch::docutils-0.21.2-pyhd8ed1ab_0 \n",
      "  importlib-metadata conda-forge/noarch::importlib-metadata-8.5.0-pyha770c72_0 \n",
      "  mock               conda-forge/noarch::mock-5.1.0-pyhd8ed1ab_0 \n",
      "  packaging          conda-forge/noarch::packaging-24.1-pyhd8ed1ab_0 \n",
      "\n",
      "The following packages will be UPDATED:\n",
      "\n",
      "  ca-certificates                       2024.7.4-hbcca054_0 --> 2024.8.30-hbcca054_0 \n",
      "  certifi                             2024.7.4-pyhd8ed1ab_0 --> 2024.8.30-pyhd8ed1ab_0 \n",
      "  importlib_metadata                       8.2.0-hd8ed1ab_0 --> 8.5.0-hd8ed1ab_0 \n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "                                                                                \n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n"
     ]
    }
   ],
   "source": [
    "!conda config --add channels conda-forge\n",
    "!conda config --add channels bioconda\n",
    "!conda install -y clustalw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d50de023",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/anaconda3/envs/python3/bin/clustalw2\r\n"
     ]
    }
   ],
   "source": [
    "!which clustalw2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95bef0bf-1c45-4f76-9c3c-4a1828883ae9",
   "metadata": {},
   "source": [
    "### Process with Clustalw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5f1e1890-9ef6-4afd-a128-9d1cac2ff90e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import datetime \n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "from Bio import AlignIO\n",
    "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor, DistanceCalculator\n",
    "\n",
    "# Define the paths\n",
    "fasta_file = \"sequences_subset.fasta\"\n",
    "clustalw_exe = \"/home/ec2-user/anaconda3/envs/python3/bin/clustalw2\"\n",
    "seq_algn_file = \"sequences_subset.aln\"\n",
    "\n",
    "start_time = datetime.datetime.now()\n",
    "print(f\"Process started at: {start_time}\")\n",
    "\n",
    "# Run ClustalW for multiple sequence alignment using subprocess\n",
    "try:\n",
    "    subprocess.run([clustalw_exe, \"-INFILE=\" + fasta_file, \"-OUTFILE=\" + seq_algn_file, \"-OUTPUT=FASTA\"], check=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(\"Error running ClustalW:\", e)\n",
    "    exit(1)\n",
    "\n",
    "end_time = datetime.datetime.now()\n",
    "print(f\"Process ended at: {end_time}\")\n",
    "\n",
    "# Calculate the duration\n",
    "duration = end_time - start_time\n",
    "print(f\"Total time taken: {duration}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c117ee61",
   "metadata": {},
   "source": [
    "## Sequence alignmnet using augur align (Nextstrain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aed1ae95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting nextstrain-cli\n",
      "  Downloading nextstrain_cli-8.5.3-py3-none-any.whl.metadata (3.8 kB)\n",
      "Requirement already satisfied: docutils in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-cli) (0.16)\n",
      "Collecting fasteners (from nextstrain-cli)\n",
      "  Downloading fasteners-0.19-py3-none-any.whl.metadata (4.9 kB)\n",
      "Requirement already satisfied: packaging in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-cli) (21.3)\n",
      "Collecting pyjwt>=2.0.0 (from pyjwt[crypto]>=2.0.0->nextstrain-cli)\n",
      "  Downloading PyJWT-2.9.0-py3-none-any.whl.metadata (3.0 kB)\n",
      "Requirement already satisfied: pyparsing>=3.0.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-cli) (3.1.2)\n",
      "Requirement already satisfied: requests in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-cli) (2.32.3)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-cli) (4.12.2)\n",
      "Collecting wcmatch>=6.0 (from nextstrain-cli)\n",
      "  Downloading wcmatch-9.0-py3-none-any.whl.metadata (4.9 kB)\n",
      "Requirement already satisfied: wrapt in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-cli) (1.16.0)\n",
      "Requirement already satisfied: fsspec!=2023.9.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-cli) (2024.6.1)\n",
      "Collecting s3fs!=2023.9.1,>=2021.04.0 (from s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli)\n",
      "  Downloading s3fs-2024.9.0-py3-none-any.whl.metadata (1.6 kB)\n",
      "Requirement already satisfied: importlib-resources>=5.3.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-cli) (6.4.0)\n",
      "Requirement already satisfied: cryptography>=3.4.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from pyjwt[crypto]>=2.0.0->nextstrain-cli) (42.0.8)\n",
      "Collecting aiobotocore<3.0.0,>=2.5.4 (from s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli)\n",
      "  Downloading aiobotocore-2.15.1-py3-none-any.whl.metadata (23 kB)\n",
      "Collecting fsspec!=2023.9.1 (from nextstrain-cli)\n",
      "  Downloading fsspec-2024.9.0-py3-none-any.whl.metadata (11 kB)\n",
      "Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (3.9.5)\n",
      "Collecting bracex>=2.1.1 (from wcmatch>=6.0->nextstrain-cli)\n",
      "  Downloading bracex-2.5-py3-none-any.whl.metadata (3.5 kB)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from requests->nextstrain-cli) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from requests->nextstrain-cli) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from requests->nextstrain-cli) (2.2.2)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from requests->nextstrain-cli) (2024.8.30)\n",
      "Requirement already satisfied: botocore<1.35.24,>=1.35.16 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (1.35.16)\n",
      "Collecting aioitertools<1.0.0,>=0.5.1 (from aiobotocore<3.0.0,>=2.5.4->s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli)\n",
      "  Downloading aioitertools-0.12.0-py3-none-any.whl.metadata (3.8 kB)\n",
      "Requirement already satisfied: boto3<1.35.24,>=1.35.16 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from aiobotocore[boto3]<3.0.0,>=2.5.4; extra == \"boto3\"->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (1.35.16)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (1.3.1)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (23.2.0)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (1.4.1)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (6.0.5)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (1.9.4)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (4.0.3)\n",
      "Requirement already satisfied: cffi>=1.12 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from cryptography>=3.4.0->pyjwt[crypto]>=2.0.0->nextstrain-cli) (1.16.0)\n",
      "Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from boto3<1.35.24,>=1.35.16->aiobotocore[boto3]<3.0.0,>=2.5.4; extra == \"boto3\"->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (1.0.1)\n",
      "Requirement already satisfied: s3transfer<0.11.0,>=0.10.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from boto3<1.35.24,>=1.35.16->aiobotocore[boto3]<3.0.0,>=2.5.4; extra == \"boto3\"->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (0.10.2)\n",
      "Requirement already satisfied: python-dateutil<3.0.0,>=2.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from botocore<1.35.24,>=1.35.16->aiobotocore<3.0.0,>=2.5.4->s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (2.9.0)\n",
      "Requirement already satisfied: pycparser in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from cffi>=1.12->cryptography>=3.4.0->pyjwt[crypto]>=2.0.0->nextstrain-cli) (2.22)\n",
      "Requirement already satisfied: six>=1.5 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from python-dateutil<3.0.0,>=2.1->botocore<1.35.24,>=1.35.16->aiobotocore<3.0.0,>=2.5.4->s3fs!=2023.9.1,>=2021.04.0->s3fs[boto3]!=2023.9.1,>=2021.04.0->nextstrain-cli) (1.16.0)\n",
      "Downloading nextstrain_cli-8.5.3-py3-none-any.whl (174 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m174.5/174.5 kB\u001b[0m \u001b[31m27.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading PyJWT-2.9.0-py3-none-any.whl (22 kB)\n",
      "Downloading s3fs-2024.9.0-py3-none-any.whl (29 kB)\n",
      "Downloading fsspec-2024.9.0-py3-none-any.whl (179 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m179.3/179.3 kB\u001b[0m \u001b[31m1.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mta \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading wcmatch-9.0-py3-none-any.whl (39 kB)\n",
      "Downloading fasteners-0.19-py3-none-any.whl (18 kB)\n",
      "Downloading aiobotocore-2.15.1-py3-none-any.whl (77 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m77.4/77.4 kB\u001b[0m \u001b[31m15.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading bracex-2.5-py3-none-any.whl (11 kB)\n",
      "Downloading aioitertools-0.12.0-py3-none-any.whl (24 kB)\n",
      "Installing collected packages: pyjwt, fsspec, fasteners, bracex, aioitertools, wcmatch, aiobotocore, s3fs, nextstrain-cli\n",
      "  Attempting uninstall: fsspec\n",
      "    Found existing installation: fsspec 2024.6.1\n",
      "    Uninstalling fsspec-2024.6.1:\n",
      "      Successfully uninstalled fsspec-2024.6.1\n",
      "  Attempting uninstall: s3fs\n",
      "    Found existing installation: s3fs 0.4.2\n",
      "    Uninstalling s3fs-0.4.2:\n",
      "      Successfully uninstalled s3fs-0.4.2\n",
      "Successfully installed aiobotocore-2.15.1 aioitertools-0.12.0 bracex-2.5 fasteners-0.19 fsspec-2024.9.0 nextstrain-cli-8.5.3 pyjwt-2.9.0 s3fs-2024.9.0 wcmatch-9.0\n"
     ]
    }
   ],
   "source": [
    "!pip install nextstrain-cli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1f3ae62d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting nextstrain-augur\n",
      "  Downloading nextstrain_augur-26.0.0-py3-none-any.whl.metadata (6.1 kB)\n",
      "Collecting bcbio-gff==0.7.*,>=0.7.1 (from nextstrain-augur)\n",
      "  Downloading bcbio_gff-0.7.1-py3-none-any.whl.metadata (343 bytes)\n",
      "Requirement already satisfied: biopython==1.*,>=1.80 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-augur) (1.84)\n",
      "Collecting cvxopt==1.*,>=1.1.9 (from nextstrain-augur)\n",
      "  Downloading cvxopt-1.3.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (1.3 kB)\n",
      "Collecting isodate==0.6.* (from nextstrain-augur)\n",
      "  Downloading isodate-0.6.1-py2.py3-none-any.whl.metadata (9.6 kB)\n",
      "Collecting jsonschema==3.*,>=3.0.0 (from nextstrain-augur)\n",
      "  Downloading jsonschema-3.2.0-py2.py3-none-any.whl.metadata (7.8 kB)\n",
      "Requirement already satisfied: networkx<4,>=2.5 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-augur) (3.3)\n",
      "Requirement already satisfied: numpy==1.* in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-augur) (1.22.4)\n",
      "Requirement already satisfied: packaging>=19.2 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-augur) (21.3)\n",
      "Collecting pandas==1.*,>=1.0.0 (from nextstrain-augur)\n",
      "  Downloading pandas-1.5.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (11 kB)\n",
      "Collecting phylo-treetime<0.12,>=0.11.2 (from nextstrain-augur)\n",
      "  Downloading phylo_treetime-0.11.4-py3-none-any.whl.metadata (13 kB)\n",
      "Collecting pyfastx<3.0,>=1.0.0 (from nextstrain-augur)\n",
      "  Downloading pyfastx-2.1.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (30 kB)\n",
      "Collecting python-calamine>=0.2.0 (from nextstrain-augur)\n",
      "  Downloading python_calamine-0.2.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.1 kB)\n",
      "Requirement already satisfied: scipy==1.* in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-augur) (1.14.0)\n",
      "Collecting xopen<3,>=1.7.0 (from xopen[zstd]<3,>=1.7.0->nextstrain-augur)\n",
      "  Downloading xopen-2.0.2-py3-none-any.whl.metadata (15 kB)\n",
      "Requirement already satisfied: importlib-resources>=5.3.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from nextstrain-augur) (6.4.0)\n",
      "Requirement already satisfied: six in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from bcbio-gff==0.7.*,>=0.7.1->nextstrain-augur) (1.16.0)\n",
      "Requirement already satisfied: attrs>=17.4.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from jsonschema==3.*,>=3.0.0->nextstrain-augur) (23.2.0)\n",
      "Requirement already satisfied: pyrsistent>=0.14.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from jsonschema==3.*,>=3.0.0->nextstrain-augur) (0.20.0)\n",
      "Requirement already satisfied: setuptools in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from jsonschema==3.*,>=3.0.0->nextstrain-augur) (71.0.4)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from pandas==1.*,>=1.0.0->nextstrain-augur) (2.9.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from pandas==1.*,>=1.0.0->nextstrain-augur) (2024.1)\n",
      "Collecting numpy==1.* (from nextstrain-augur)\n",
      "  Downloading numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m61.0/61.0 kB\u001b[0m \u001b[31m11.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from packaging>=19.2->nextstrain-augur) (3.1.2)\n",
      "Requirement already satisfied: matplotlib>=2.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from phylo-treetime<0.12,>=0.11.2->nextstrain-augur) (3.8.4)\n",
      "Collecting isal>=1.6.1 (from xopen<3,>=1.7.0->xopen[zstd]<3,>=1.7.0->nextstrain-augur)\n",
      "  Downloading isal-1.7.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (10 kB)\n",
      "Collecting zlib-ng>=0.4.1 (from xopen<3,>=1.7.0->xopen[zstd]<3,>=1.7.0->nextstrain-augur)\n",
      "  Downloading zlib_ng-0.5.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.7 kB)\n",
      "Requirement already satisfied: zstandard<1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from xopen[zstd]<3,>=1.7.0->nextstrain-augur) (0.23.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib>=2.0->phylo-treetime<0.12,>=0.11.2->nextstrain-augur) (1.2.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib>=2.0->phylo-treetime<0.12,>=0.11.2->nextstrain-augur) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib>=2.0->phylo-treetime<0.12,>=0.11.2->nextstrain-augur) (4.53.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib>=2.0->phylo-treetime<0.12,>=0.11.2->nextstrain-augur) (1.4.5)\n",
      "Requirement already satisfied: pillow>=8 in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from matplotlib>=2.0->phylo-treetime<0.12,>=0.11.2->nextstrain-augur) (10.4.0)\n",
      "Downloading nextstrain_augur-26.0.0-py3-none-any.whl (340 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m340.9/340.9 kB\u001b[0m \u001b[31m5.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m\n",
      "\u001b[?25hDownloading bcbio_gff-0.7.1-py3-none-any.whl (16 kB)\n",
      "Downloading cvxopt-1.3.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.6 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.6/13.6 MB\u001b[0m \u001b[31m57.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading isodate-0.6.1-py2.py3-none-any.whl (41 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m41.7/41.7 kB\u001b[0m \u001b[31m7.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.3/56.3 kB\u001b[0m \u001b[31m8.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading pandas-1.5.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.1 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m12.1/12.1 MB\u001b[0m \u001b[31m104.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m0:01\u001b[0m\n",
      "\u001b[?25hDownloading numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.2 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m18.2/18.2 MB\u001b[0m \u001b[31m85.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading phylo_treetime-0.11.4-py3-none-any.whl (148 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m148.4/148.4 kB\u001b[0m \u001b[31m22.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading pyfastx-2.1.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.1 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.1/1.1 MB\u001b[0m \u001b[31m14.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m\n",
      "\u001b[?25hDownloading python_calamine-0.2.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (819 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m819.5/819.5 kB\u001b[0m \u001b[31m17.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading xopen-2.0.2-py3-none-any.whl (17 kB)\n",
      "Downloading isal-1.7.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (258 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m259.0/259.0 kB\u001b[0m \u001b[31m4.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m\n",
      "\u001b[?25hDownloading zlib_ng-0.5.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (108 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m108.2/108.2 kB\u001b[0m \u001b[31m2.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mta \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: pyfastx, zlib-ng, python-calamine, numpy, jsonschema, isodate, isal, cvxopt, xopen, pandas, bcbio-gff, phylo-treetime, nextstrain-augur\n",
      "  Attempting uninstall: numpy\n",
      "    Found existing installation: numpy 1.22.4\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Uninstalling numpy-1.22.4:\n",
      "      Successfully uninstalled numpy-1.22.4\n",
      "  Attempting uninstall: jsonschema\n",
      "    Found existing installation: jsonschema 4.23.0\n",
      "    Uninstalling jsonschema-4.23.0:\n",
      "      Successfully uninstalled jsonschema-4.23.0\n",
      "  Attempting uninstall: pandas\n",
      "    Found existing installation: pandas 2.2.2\n",
      "    Uninstalling pandas-2.2.2:\n",
      "      Successfully uninstalled pandas-2.2.2\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "mkl-fft 1.3.10 requires mkl, which is not installed.\n",
      "dask-expr 1.1.9 requires pandas>=2, but you have pandas 1.5.3 which is incompatible.\n",
      "jupyter-events 0.10.0 requires jsonschema[format-nongpl]>=4.18.0, but you have jsonschema 3.2.0 which is incompatible.\n",
      "jupyter-server 2.14.2 requires packaging>=22.0, but you have packaging 21.3 which is incompatible.\n",
      "jupyterlab-server 2.27.3 requires jsonschema>=4.18.0, but you have jsonschema 3.2.0 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0mSuccessfully installed bcbio-gff-0.7.1 cvxopt-1.3.2 isal-1.7.0 isodate-0.6.1 jsonschema-3.2.0 nextstrain-augur-26.0.0 numpy-1.26.4 pandas-1.5.3 phylo-treetime-0.11.4 pyfastx-2.1.0 python-calamine-0.2.3 xopen-2.0.2 zlib-ng-0.5.0\n"
     ]
    }
   ],
   "source": [
    "!pip install nextstrain-augur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ed8d5339",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 23.3.1\n",
      "  latest version: 24.7.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=24.7.1\n",
      "\n",
      "\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/ec2-user/anaconda3/envs/python3\n",
      "\n",
      "  added / updated specs:\n",
      "    - fasttree\n",
      "    - mafft\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    fasttree-2.1.11            |       h031d066_4         261 KB  bioconda\n",
      "    mafft-7.505                |       hec16e2b_0         3.5 MB  bioconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:         3.7 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  fasttree           bioconda/linux-64::fasttree-2.1.11-h031d066_4 \n",
      "  mafft              bioconda/linux-64::mafft-7.505-hec16e2b_0 \n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "fasttree-2.1.11      | 261 KB    |                                       |   0% \n",
      "mafft-7.505          | 3.5 MB    |                                       |   0% \u001b[A\n",
      "fasttree-2.1.11      | 261 KB    | ##################################### | 100% \u001b[A\n",
      "                                                                                \u001b[A\n",
      "                                                                                \u001b[A\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n"
     ]
    }
   ],
   "source": [
    "!conda install -c bioconda mafft fasttree -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "51ac3e60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      "using mafft to align via:\r\n",
      "\tmafft --reorder --anysymbol --nomemsave --adjustdirection --thread 1 aligned_subset_augur.fasta.to_align.fasta 1> aligned_subset_augur.fasta 2> aligned_subset_augur.fasta.log \r\n",
      "\r\n",
      "\tKatoh et al, Nucleic Acid Research, vol 30, issue 14\r\n",
      "\thttps://doi.org/10.1093%2Fnar%2Fgkf436\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!augur align --sequences sequences_subset.fasta --output aligned_subset_augur.fasta --fill-gaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0c6e404b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/SageMaker/nosi-phylogenetic-tree\n"
     ]
    }
   ],
   "source": [
    "%cd ../.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bb051710",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/ec2-user/SageMaker/nosi-phylogenetic-tree'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd93b0e8-5b6e-4666-a5d9-209a78275e98",
   "metadata": {},
   "source": [
    "This process aligns the SARS-CoV-2 sequences, preparing them for phylogenetic tree construction."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62a23c0b-a22c-4300-881a-0b43de5566b4",
   "metadata": {},
   "source": [
    "## 3.2 Manage Computational Intensity through Cloud Computing\n",
    "Due to the large size of metagenomic datasets, sequence alignment can be computationally intensive. Utilizing cloud computing resources can significantly enhance the efficiency and speed of these tasks.\n",
    "\n",
    "**Benefits of Cloud Computing for Sequence Alignment:**\n",
    "- Scalability: Easily scale up resources based on the demand of the computation.\n",
    "- Cost-Effectiveness: Pay-as-you-go models allow for cost savings by only using resources when needed.\n",
    "- Accessibility: Access computational resources and data from anywhere, facilitating collaboration among researchers."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd904c45-c2bc-486e-873d-a282dd7df8db",
   "metadata": {},
   "source": [
    "## 3.3 Phylogenetic Tree Reconstruction using USHER\n",
    "USHER (Ultrafast Sample Placement on Existing tRee) is a tool designed to place samples on a given phylogenetic tree rapidly. It is beneficial for large-scale phylogenetic analysis and real-time epidemiology.\n",
    "\n",
    "### Steps to Use USHER for Phylogenetic Tree Reconstruction:\n",
    "1. Clone USHER Repository:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "192636aa-9bb7-445b-8fdb-0a495638d2d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cloning into 'usher'...\n",
      "remote: Enumerating objects: 11290, done.\u001b[K\n",
      "remote: Counting objects: 100% (1245/1245), done.\u001b[K\n",
      "remote: Compressing objects: 100% (529/529), done.\u001b[K\n",
      "remote: Total 11290 (delta 798), reused 1030 (delta 694), pack-reused 10045 (from 1)\u001b[K\n",
      "Receiving objects: 100% (11290/11290), 40.04 MiB | 38.61 MiB/s, done.\n",
      "Resolving deltas: 100% (7943/7943), done.\n"
     ]
    }
   ],
   "source": [
    "!git clone https://github.com/yatisht/usher.git"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07dc6a0a-4b69-4204-a53a-1b7f4d8d6ce6",
   "metadata": {},
   "source": [
    "2. Installing Dependencies:\n",
    "- Update the conda environment with the necessary dependencies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "148b9b09-c05b-4ad5-b492-dca35577ef1c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!conda env update -f usher/workflows/envs/usher.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ae46892-4d47-41ed-baea-da68e319b403",
   "metadata": {},
   "source": [
    "3.\tInstalling Additional Packages:\n",
    "- Install the required packages mafft and fasttree:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d93be479-c1b1-4d85-a8f0-c231d806f80f",
   "metadata": {},
   "outputs": [],
   "source": [
    "!conda install -c bioconda mafft fasttree -y"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aacaad5e-33e7-4a2d-aa3c-b135b4294e32",
   "metadata": {},
   "source": [
    "4.\tAligning Sequences:\n",
    "- Use mafft to align your sequences and output them to aligned_sequences.fasta:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "52fbb44e-4a6f-49d8-9178-a6c43cecdd81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 10240 kb\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 3961 ambiguous characters.\n",
      "    1 / 10\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      "    0 / 10\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP     3 / 9 \n",
      "Reallocating..done. *alloclen = 60816\n",
      "STEP     9 / 9 \n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      "    0 / 10\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 10\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP     9 / 9 \n",
      "done.\n",
      "\n",
      "disttbfast (nuc) Version 7.505\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-2 (Fast but rough)\n",
      " Progressive method (guide trees were built 2 times.)\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!mafft --auto data/cov/sequences_subset.fasta > data/cov/aligned_sequences_mafft_subset.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6c57f91-4169-4d84-800e-e89d8e871d23",
   "metadata": {},
   "source": [
    "5.\tGenerating VCF File:\n",
    "- Convert the aligned sequences to a VCF file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8ef45d4e-0a71-4a73-8fb0-8563a1e6a788",
   "metadata": {},
   "outputs": [],
   "source": [
    "!faToVcf aligned_sequences_subset.fasta seq_subset.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c2dd6b8-a371-4951-a0a4-b372b4bc08b5",
   "metadata": {},
   "source": [
    "6.\tCreating Newick Tree File:\n",
    "- Use fasttree to generate a Newick tree file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d025cf79-e7a2-4e85-b2b0-0afb59a4270f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastTree Version 2.1.11 Double precision (No SSE3)\n",
      "Alignment: aligned_sequences.fasta\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: SH-like 1000\n",
      "Search: Normal +NNI +SPR (2 rounds range 10) +ML-NNI opt-each=1\n",
      "TopHits: 1.00*sqrtN close=default refresh=0.80\n",
      "ML Model: Jukes-Cantor, CAT approximation with 20 rate categories\n",
      "Ignored unknown character k (seen 10 times)\n",
      "Ignored unknown character m (seen 4 times)\n",
      "Ignored unknown character n (seen 115807 times)\n",
      "Ignored unknown character r (seen 25 times)\n",
      "Ignored unknown character s (seen 2 times)\n",
      "Ignored unknown character w (seen 3 times)\n",
      "Ignored unknown character y (seen 41 times)\n",
      "Initial topology in 1.64 seconds0 of    162   165 seqs (at seed    100)   \n",
      "Refining topology: 29 rounds ME-NNIs, 2 rounds ME-SPRs, 15 rounds ML-NNIs\n",
      "Total branch-length 0.067 after 10.73 sec, 1 of 163 splits   8 changes (max delta 0.000)    \n",
      "\n",
      "WARNING! This alignment consists of closely-related and very-long sequences.\n",
      "WARNING! FastTree (or other standard maximum-likelihood tools)\n",
      "may not be appropriate for aligments of very closely-related sequences\n",
      "like this one, as FastTree does not account for recombination or gene conversion\n",
      "\n",
      "ML-NNI round 1: LogLk = -62797.777 NNIs 37 max delta 14.59 Time 16.09es (max delta 7.176)   \n",
      "Switched to using 20 rate categories (CAT approximation)19 of 20   \n",
      "Rate categories were divided by 0.642 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "Use -gamma for approximate but comparable Gamma(20) log-likelihoods\n",
      "ML-NNI round 2: LogLk = -61603.107 NNIs 19 max delta 6.64 Time 19.42es (max delta 3.571)   \n",
      "ML-NNI round 3: LogLk = -61592.525 NNIs 9 max delta 6.64 Time 20.75ges (max delta 6.644)   \n",
      "ML-NNI round 4: LogLk = -61586.959 NNIs 7 max delta 5.57 Time 21.36\n",
      "ML-NNI round 5: LogLk = -61586.941 NNIs 2 max delta 0.00 Time 21.81\n",
      "Turning off heuristics for final round of ML NNIs (converged)\n",
      "ML-NNI round 6: LogLk = -61586.939 NNIs 0 max delta 0.00 Time 25.44 (final)\n",
      "Optimize all lengths: LogLk = -61586.939 Time 26.38\n",
      "Total time: 36.30 seconds Unique: 165/165 Bad splits: 0/162ernal splits   \n"
     ]
    }
   ],
   "source": [
    "!fasttree -nt aligned_sequences_subset.fasta > reference_sequences_subset.nwk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0a0e0e8-0a8f-490b-941f-2a59860ec5a6",
   "metadata": {},
   "source": [
    "7.\tRunning USHER:\n",
    "- With the aligned sequences, VCF file, and Newick tree file, run USHER:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3fbc47b6-33b4-49c1-bef9-c986abdcc3c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing 8 worker threads.\n",
      "\n",
      "Loading input tree.\n",
      "Completed in 3 msec \n",
      "\n",
      "Loading VCF file.\n",
      "Completed in 1 msec \n",
      "\n",
      "Computing parsimonious assignments for input variants.\n",
      "At variant site 8\n",
      "At variant site 9\n",
      "At variant site 10\n",
      "At variant site 11\n",
      "At variant site 12\n",
      "At variant site 14\n",
      "At variant site 15\n",
      "At variant site 16\n",
      "At variant site 17\n",
      "At variant site 18\n",
      "At variant site 20\n",
      "At variant site 19\n",
      "At variant site 21\n",
      "At variant site 22\n",
      "At variant site 23\n",
      "At variant site 25\n",
      "At variant site 24\n",
      "At variant site 27\n",
      "At variant site 26\n",
      "At variant site 28\n",
      "At variant site 29\n",
      "At variant site 30\n",
      "At variant site 31\n",
      "At variant site 34\n",
      "At variant site 35\n",
      "At variant site 36\n",
      "At variant site 37\n",
      "At variant site 44\n",
      "At variant site 66\n",
      "At variant site 72\n",
      "At variant site 105\n",
      "At variant site 127\n",
      "At variant site 140\n",
      "At variant site 143\n",
      "At variant site 174\n",
      "At variant site 188\n",
      "At variant site 193\n",
      "At variant site 201\n",
      "At variant site 203\n",
      "At variant site 204\n",
      "At variant site 210\n",
      "At variant site 217\n",
      "At variant site 218\n",
      "At variant site 222\n",
      "At variant site 225\n",
      "At variant site 241\n",
      "At variant site 261\n",
      "At variant site 370\n",
      "At variant site 385\n",
      "At variant site 405\n",
      "At variant site 407\n",
      "At variant site 410\n",
      "At variant site 445\n",
      "At variant site 449\n",
      "At variant site 490\n",
      "At variant site 510\n",
      "At variant site 542\n",
      "At variant site 632\n",
      "At variant site 654\n",
      "At variant site 670\n",
      "At variant site 679\n",
      "At variant site 733\n",
      "At variant site 756\n",
      "At variant site 806\n",
      "At variant site 823\n",
      "At variant site 832\n",
      "At variant site 840\n",
      "At variant site 859\n",
      "At variant site 875\n",
      "At variant site 897\n",
      "At variant site 900\n",
      "At variant site 906\n",
      "At variant site 913\n",
      "At variant site 916\n",
      "At variant site 923\n",
      "At variant site 925\n",
      "At variant site 959\n",
      "At variant site 960\n",
      "At variant site 1048\n",
      "At variant site 1059\n",
      "At variant site 1085\n",
      "At variant site 1113\n",
      "At variant site 1124\n",
      "At variant site 1163\n",
      "At variant site 1179\n",
      "At variant site 1188\n",
      "At variant site 1191\n",
      "At variant site 1255\n",
      "At variant site 1267\n",
      "At variant site 1274\n",
      "At variant site 1321\n",
      "At variant site 1350\n",
      "At variant site 1390\n",
      "At variant site 1403\n",
      "At variant site 1411\n",
      "At variant site 1420\n",
      "At variant site 1427\n",
      "At variant site 1437\n",
      "At variant site 1471\n",
      "At variant site 1495\n",
      "At variant site 1498\n",
      "At variant site 1513\n",
      "At variant site 1514\n",
      "At variant site 1524\n",
      "At variant site 1545\n",
      "At variant site 1572\n",
      "At variant site 1585\n",
      "At variant site 1627\n",
      "At variant site 1646\n",
      "At variant site 1651\n",
      "At variant site 1652\n",
      "At variant site 1684\n",
      "At variant site 1685\n",
      "At variant site 1722\n",
      "At variant site 1727\n",
      "At variant site 1738\n",
      "At variant site 1752\n",
      "At variant site 1758\n",
      "At variant site 1807\n",
      "At variant site 1820\n",
      "At variant site 1825\n",
      "At variant site 1840\n",
      "At variant site 1859\n",
      "At variant site 1885\n",
      "At variant site 1887\n",
      "At variant site 1889\n",
      "At variant site 1912\n",
      "At variant site 1931\n",
      "At variant site 1968\n",
      "At variant site 1973\n",
      "At variant site 1982\n",
      "At variant site 2005\n",
      "At variant site 2007\n",
      "At variant site 2013\n",
      "At variant site 2015\n",
      "At variant site 2019\n",
      "At variant site 2043\n",
      "At variant site 2061\n",
      "At variant site 2068\n",
      "At variant site 2072\n",
      "At variant site 2075\n",
      "At variant site 2083\n",
      "At variant site 2110\n",
      "At variant site 2127\n",
      "At variant site 2128\n",
      "At variant site 2147\n",
      "At variant site 2163\n",
      "At variant site 2210\n",
      "At variant site 2218\n",
      "At variant site 2227\n",
      "At variant site 2255\n",
      "At variant site 2276\n",
      "At variant site 2284\n",
      "At variant site 2305\n",
      "At variant site 2334\n",
      "At variant site 2336\n",
      "At variant site 2395\n",
      "At variant site 2433\n",
      "At variant site 2434\n",
      "At variant site 2445\n",
      "At variant site 2453\n",
      "At variant site 2470\n",
      "At variant site 2527\n",
      "At variant site 2597\n",
      "At variant site 2506\n",
      "At variant site 2644\n",
      "At variant site 2659\n",
      "At variant site 2668\n",
      "At variant site 2675\n",
      "At variant site 2692\n",
      "At variant site 2695\n",
      "At variant site 2749\n",
      "At variant site 2790\n",
      "At variant site 2795\n",
      "At variant site 2832\n",
      "At variant site 2881\n",
      "At variant site 2910\n",
      "At variant site 2954\n",
      "At variant site 3037\n",
      "At variant site 3046\n",
      "At variant site 3058\n",
      "At variant site 3079\n",
      "At variant site 3013\n",
      "At variant site 3092\n",
      "At variant site 3096\n",
      "At variant site 3114\n",
      "At variant site 3130\n",
      "At variant site 3140\n",
      "At variant site 3177\n",
      "At variant site 3206\n",
      "At variant site 3210\n",
      "At variant site 3241\n",
      "At variant site 3242\n",
      "At variant site 3247\n",
      "At variant site 3250\n",
      "At variant site 3256\n",
      "At variant site 3267\n",
      "At variant site 3294\n",
      "At variant site 3365\n",
      "At variant site 3391\n",
      "At variant site 3411\n",
      "At variant site 3428\n",
      "At variant site 3431\n",
      "At variant site 3444\n",
      "At variant site 3448\n",
      "At variant site 3457\n",
      "At variant site 3478\n",
      "At variant site 3564\n",
      "At variant site 3567\n",
      "At variant site 3602\n",
      "At variant site 3688\n",
      "At variant site 3619\n",
      "At variant site 3701\n",
      "At variant site 3737\n",
      "At variant site 3787\n",
      "At variant site 3796\n",
      "At variant site 3817\n",
      "At variant site 3828\n",
      "At variant site 3839\n",
      "At variant site 3857\n",
      "At variant site 3873\n",
      "At variant site 3877\n",
      "At variant site 3884\n",
      "At variant site 3880\n",
      "At variant site 3923\n",
      "At variant site 3927\n",
      "At variant site 3952\n",
      "At variant site 3961\n",
      "At variant site 3905\n",
      "At variant site 3994\n",
      "At variant site 4002\n",
      "At variant site 4009\n",
      "At variant site 4071\n",
      "At variant site 4079\n",
      "At variant site 4036\n",
      "At variant site 4090\n",
      "At variant site 3974\n",
      "At variant site 4106\n",
      "At variant site 4109\n",
      "At variant site 4158\n",
      "At variant site 4181\n",
      "At variant site 4184\n",
      "At variant site 4234\n",
      "At variant site 4236\n",
      "At variant site 4237\n",
      "At variant site 4255\n",
      "At variant site 4276\n",
      "At variant site 4285\n",
      "At variant site 4321\n",
      "At variant site 4282\n",
      "At variant site 4402\n",
      "At variant site 4455\n",
      "At variant site 4456\n",
      "At variant site 4475\n",
      "At variant site 4514\n",
      "At variant site 4516\n",
      "At variant site 4543\n",
      "At variant site 4551\n",
      "At variant site 4579\n",
      "At variant site 4586\n",
      "At variant site 4684\n",
      "At variant site 4686\n",
      "At variant site 4755\n",
      "At variant site 4784\n",
      "At variant site 4795\n",
      "At variant site 4829\n",
      "At variant site 4876\n",
      "At variant site 4878\n",
      "At variant site 4893\n",
      "At variant site 4916\n",
      "At variant site 4918\n",
      "At variant site 4921\n",
      "At variant site 4926\n",
      "At variant site 4931\n",
      "At variant site 4965\n",
      "At variant site 4999\n",
      "At variant site 5011\n",
      "At variant site 5164\n",
      "At variant site 5182\n",
      "At variant site 5183\n",
      "At variant site 4927\n",
      "At variant site 5184\n",
      "At variant site 5192\n",
      "At variant site 5230\n",
      "At variant site 5235\n",
      "At variant site 5281\n",
      "At variant site 5301\n",
      "At variant site 5284\n",
      "At variant site 5314\n",
      "At variant site 5386\n",
      "At variant site 5388\n",
      "At variant site 5406\n",
      "At variant site 5428\n",
      "At variant site 5437\n",
      "At variant site 5470\n",
      "At variant site 5502\n",
      "At variant site 5504\n",
      "At variant site 5518\n",
      "At variant site 5544\n",
      "At variant site 5555\n",
      "At variant site 5575\n",
      "At variant site 5577\n",
      "At variant site 5584\n",
      "At variant site 5629\n",
      "At variant site 5641\n",
      "At variant site 5646\n",
      "At variant site 5648\n",
      "At variant site 5654\n",
      "At variant site 5720\n",
      "At variant site 5835\n",
      "At variant site 5869\n",
      "At variant site 5924\n",
      "At variant site 5926\n",
      "At variant site 5931\n",
      "At variant site 5944\n",
      "At variant site 5947\n",
      "At variant site 5950\n",
      "At variant site 5974\n",
      "At variant site 5986\n",
      "At variant site 6023\n",
      "At variant site 6037\n",
      "At variant site 6040\n",
      "At variant site 6052\n",
      "At variant site 6004\n",
      "At variant site 6085\n",
      "At variant site 6101\n",
      "At variant site 6183\n",
      "At variant site 6191\n",
      "At variant site 6196\n",
      "At variant site 6199\n",
      "At variant site 6226\n",
      "At variant site 6250\n",
      "At variant site 6285\n",
      "At variant site 6286\n",
      "At variant site 6317\n",
      "At variant site 6319\n",
      "At variant site 6325\n",
      "At variant site 6352\n",
      "At variant site 6402\n",
      "At variant site 6404\n",
      "At variant site 6408\n",
      "At variant site 6443\n",
      "At variant site 6472\n",
      "At variant site 6476\n",
      "At variant site 6486\n",
      "At variant site 6503\n",
      "At variant site 6517\n",
      "At variant site 6525\n",
      "At variant site 6536\n",
      "At variant site 6538\n",
      "At variant site 6539\n",
      "At variant site 6541\n",
      "At variant site 6576\n",
      "At variant site 6589\n",
      "At variant site 6613\n",
      "At variant site 6618\n",
      "At variant site 6627\n",
      "At variant site 6633\n",
      "At variant site 6636\n",
      "At variant site 6730\n",
      "At variant site 6752\n",
      "At variant site 6781\n",
      "At variant site 6889\n",
      "At variant site 6896\n",
      "At variant site 6770\n",
      "At variant site 6982\n",
      "At variant site 6786\n",
      "At variant site 6886\n",
      "At variant site 7056\n",
      "At variant site 6954\n",
      "At variant site 7086\n",
      "At variant site 7042\n",
      "At variant site 7135\n",
      "At variant site 7068\n",
      "At variant site 7091\n",
      "At variant site 7124\n",
      "At variant site 7142\n",
      "At variant site 7201\n",
      "At variant site 7300\n",
      "At variant site 7344\n",
      "At variant site 7420\n",
      "At variant site 7424\n",
      "At variant site 7540\n",
      "At variant site 7262\n",
      "At variant site 7564\n",
      "At variant site 7604\n",
      "At variant site 7618\n",
      "At variant site 7267\n",
      "At variant site 7767\n",
      "At variant site 7791\n",
      "At variant site 7798\n",
      "At variant site 7842\n",
      "At variant site 7843\n",
      "At variant site 7765\n",
      "At variant site 7851\n",
      "At variant site 7855\n",
      "At variant site 7926\n",
      "At variant site 8016\n",
      "At variant site 8031\n",
      "At variant site 8083\n",
      "At variant site 8090\n",
      "At variant site 8110\n",
      "At variant site 8139\n",
      "At variant site 8162\n",
      "At variant site 8167\n",
      "At variant site 8170\n",
      "At variant site 8208\n",
      "At variant site 8240\n",
      "At variant site 8293\n",
      "At variant site 8323\n",
      "At variant site 8393\n",
      "At variant site 8407\n",
      "At variant site 8449\n",
      "At variant site 8593\n",
      "At variant site 8594\n",
      "At variant site 8595\n",
      "At variant site 8668\n",
      "At variant site 8692\n",
      "At variant site 8723\n",
      "At variant site 8737\n",
      "At variant site 8764\n",
      "At variant site 8782\n",
      "At variant site 8809\n",
      "At variant site 8834\n",
      "At variant site 8835\n",
      "At variant site 8844\n",
      "At variant site 8872\n",
      "At variant site 8889\n",
      "At variant site 8917\n",
      "At variant site 8947\n",
      "At variant site 8964\n",
      "At variant site 8980\n",
      "At variant site 8986\n",
      "At variant site 8991\n",
      "At variant site 9053\n",
      "At variant site 9072\n",
      "At variant site 9073\n",
      "At variant site 9085\n",
      "At variant site 9142\n",
      "At variant site 9152\n",
      "At variant site 9091\n",
      "At variant site 9165\n",
      "At variant site 9166\n",
      "At variant site 9190\n",
      "At variant site 9195\n",
      "At variant site 9203\n",
      "At variant site 9211\n",
      "At variant site 9241\n",
      "At variant site 9286\n",
      "At variant site 9332\n",
      "At variant site 9344\n",
      "At variant site 9220\n",
      "At variant site 9424\n",
      "At variant site 9430\n",
      "At variant site 9435\n",
      "At variant site 9419\n",
      "At variant site 9526\n",
      "At variant site 9532\n",
      "At variant site 9534\n",
      "At variant site 9544\n",
      "At variant site 9561\n",
      "At variant site 9565\n",
      "At variant site 9611\n",
      "At variant site 9583\n",
      "At variant site 9614\n",
      "At variant site 9693\n",
      "At variant site 9738\n",
      "At variant site 9745\n",
      "At variant site 9750\n",
      "At variant site 9754\n",
      "At variant site 9848\n",
      "At variant site 9852\n",
      "At variant site 9857\n",
      "At variant site 9866\n",
      "At variant site 9867\n",
      "At variant site 9868\n",
      "At variant site 9891\n",
      "At variant site 9892\n",
      "At variant site 9893\n",
      "At variant site 9952\n",
      "At variant site 9970\n",
      "At variant site 10029\n",
      "At variant site 10036\n",
      "At variant site 10097\n",
      "At variant site 10116\n",
      "At variant site 10117\n",
      "At variant site 10141\n",
      "At variant site 10183\n",
      "At variant site 10189\n",
      "At variant site 10198\n",
      "At variant site 10204\n",
      "At variant site 10273\n",
      "At variant site 10279\n",
      "At variant site 10310\n",
      "At variant site 10301\n",
      "At variant site 10323\n",
      "At variant site 10317\n",
      "At variant site 10335\n",
      "At variant site 10319\n",
      "At variant site 10345\n",
      "At variant site 10364\n",
      "At variant site 10437\n",
      "At variant site 10447\n",
      "At variant site 10449\n",
      "At variant site 10471\n",
      "At variant site 10369\n",
      "At variant site 10537\n",
      "At variant site 10543\n",
      "At variant site 10591\n",
      "At variant site 10626\n",
      "At variant site 10631\n",
      "At variant site 10702\n",
      "At variant site 10714\n",
      "At variant site 10756\n",
      "At variant site 10775\n",
      "At variant site 10798\n",
      "At variant site 10861\n",
      "At variant site 10908\n",
      "At variant site 10954\n",
      "At variant site 10977\n",
      "At variant site 10985\n",
      "At variant site 11020\n",
      "At variant site 11042\n",
      "At variant site 11074\n",
      "At variant site 11083\n",
      "At variant site 11095\n",
      "At variant site 11109\n",
      "At variant site 11117\n",
      "At variant site 11173\n",
      "At variant site 11199\n",
      "At variant site 11200\n",
      "At variant site 11201\n",
      "At variant site 11208\n",
      "At variant site 11235\n",
      "At variant site 11236\n",
      "At variant site 11299\n",
      "At variant site 11323\n",
      "At variant site 11287\n",
      "At variant site 11336\n",
      "At variant site 11409\n",
      "At variant site 11442\n",
      "At variant site 11446\n",
      "At variant site 11447\n",
      "At variant site 11451\n",
      "At variant site 11455\n",
      "At variant site 11488\n",
      "At variant site 11505\n",
      "At variant site 11513\n",
      "At variant site 11528\n",
      "At variant site 11564\n",
      "At variant site 11530\n",
      "At variant site 11611\n",
      "At variant site 11638\n",
      "At variant site 11660\n",
      "At variant site 11665\n",
      "At variant site 11695\n",
      "At variant site 11738\n",
      "At variant site 11741\n",
      "At variant site 11743\n",
      "At variant site 11767\n",
      "At variant site 11826\n",
      "At variant site 11939\n",
      "At variant site 11947\n",
      "At variant site 11999\n",
      "At variant site 12011\n",
      "At variant site 12021\n",
      "At variant site 12029\n",
      "At variant site 12040\n",
      "At variant site 12044\n",
      "At variant site 12061\n",
      "At variant site 12064\n",
      "At variant site 12067\n",
      "At variant site 12075\n",
      "At variant site 12091\n",
      "At variant site 12151\n",
      "At variant site 12154\n",
      "At variant site 12217\n",
      "At variant site 12247\n",
      "At variant site 12301\n",
      "At variant site 12329\n",
      "At variant site 12388\n",
      "At variant site 12430\n",
      "At variant site 12435\n",
      "At variant site 12450\n",
      "At variant site 12464\n",
      "At variant site 12469\n",
      "At variant site 12504\n",
      "At variant site 12516\n",
      "At variant site 12525\n",
      "At variant site 12582\n",
      "At variant site 12617\n",
      "At variant site 12632\n",
      "At variant site 12642\n",
      "At variant site 12458\n",
      "At variant site 12721\n",
      "At variant site 12738\n",
      "At variant site 12741\n",
      "At variant site 12744\n",
      "At variant site 12759\n",
      "At variant site 12750\n",
      "At variant site 12769\n",
      "At variant site 12775\n",
      "At variant site 12780\n",
      "At variant site 12837\n",
      "At variant site 12845\n",
      "At variant site 12806\n",
      "At variant site 12869\n",
      "At variant site 12871\n",
      "At variant site 12906\n",
      "At variant site 12937\n",
      "At variant site 13002\n",
      "At variant site 13003\n",
      "At variant site 13010\n",
      "At variant site 13026\n",
      "At variant site 12961\n",
      "At variant site 13048\n",
      "At variant site 13036\n",
      "At variant site 13063\n",
      "At variant site 13096\n",
      "At variant site 13117\n",
      "At variant site 13131\n",
      "At variant site 13186\n",
      "At variant site 13201\n",
      "At variant site 13320\n",
      "At variant site 13330\n",
      "At variant site 13372\n",
      "At variant site 13396\n",
      "At variant site 13414\n",
      "At variant site 13415\n",
      "At variant site 13417\n",
      "At variant site 13449\n",
      "At variant site 13473\n",
      "At variant site 13527\n",
      "At variant site 13608\n",
      "At variant site 13618\n",
      "At variant site 13658\n",
      "At variant site 13671\n",
      "At variant site 13685\n",
      "At variant site 13704\n",
      "At variant site 13706\n",
      "At variant site 13712\n",
      "At variant site 13758\n",
      "At variant site 13762\n",
      "At variant site 13851\n",
      "At variant site 13884\n",
      "At variant site 13936\n",
      "At variant site 13984\n",
      "At variant site 14005\n",
      "At variant site 14058\n",
      "At variant site 14075\n",
      "At variant site 14076\n",
      "At variant site 14111\n",
      "At variant site 14113\n",
      "At variant site 14110\n",
      "At variant site 14169\n",
      "At variant site 14178\n",
      "At variant site 14180\n",
      "At variant site 14175\n",
      "At variant site 14195\n",
      "At variant site 14220\n",
      "At variant site 14248\n",
      "At variant site 14259\n",
      "At variant site 14334\n",
      "At variant site 14286\n",
      "At variant site 14398\n",
      "At variant site 14399\n",
      "At variant site 14353\n",
      "At variant site 14469\n",
      "At variant site 14590\n",
      "At variant site 14613\n",
      "At variant site 14625\n",
      "At variant site 14628\n",
      "At variant site 14629\n",
      "At variant site 14667\n",
      "At variant site 14694\n",
      "At variant site 14776\n",
      "At variant site 14781\n",
      "At variant site 14733\n",
      "At variant site 14796\n",
      "At variant site 14847\n",
      "At variant site 14865\n",
      "At variant site 14928\n",
      "At variant site 14936\n",
      "At variant site 14964\n",
      "At variant site 14970\n",
      "At variant site 14981\n",
      "At variant site 15000\n",
      "At variant site 15003\n",
      "At variant site 15017\n",
      "At variant site 15021\n",
      "At variant site 15087\n",
      "At variant site 15228\n",
      "At variant site 15234\n",
      "At variant site 15012\n",
      "At variant site 15244\n",
      "At variant site 15255\n",
      "At variant site 15242\n",
      "At variant site 15270\n",
      "At variant site 15231\n",
      "At variant site 15258\n",
      "At variant site 15359\n",
      "At variant site 15363\n",
      "At variant site 15371\n",
      "At variant site 15399\n",
      "At variant site 15441\n",
      "At variant site 15442\n",
      "At variant site 15501\n",
      "At variant site 15512\n",
      "At variant site 15520\n",
      "At variant site 15531\n",
      "At variant site 15588\n",
      "At variant site 15645\n",
      "At variant site 15647\n",
      "At variant site 15705\n",
      "At variant site 15711\n",
      "At variant site 15729\n",
      "At variant site 15735\n",
      "At variant site 15746\n",
      "At variant site 15747\n",
      "At variant site 15751\n",
      "At variant site 15754\n",
      "At variant site 15757\n",
      "At variant site 15849\n",
      "At variant site 15897\n",
      "At variant site 15839\n",
      "At variant site 15930\n",
      "At variant site 15942\n",
      "At variant site 15943\n",
      "At variant site 15951\n",
      "At variant site 16066\n",
      "At variant site 16164\n",
      "At variant site 16167\n",
      "At variant site 16185\n",
      "At variant site 16204\n",
      "At variant site 16251\n",
      "At variant site 16272\n",
      "At variant site 16302\n",
      "At variant site 16333\n",
      "At variant site 16353\n",
      "At variant site 16364\n",
      "At variant site 16382\n",
      "At variant site 16385\n",
      "At variant site 16457\n",
      "At variant site 16491\n",
      "At variant site 16566\n",
      "At variant site 16607\n",
      "At variant site 16638\n",
      "At variant site 16671\n",
      "At variant site 16698\n",
      "At variant site 16716\n",
      "At variant site 16741\n",
      "At variant site 16742\n",
      "At variant site 16843\n",
      "At variant site 16847\n",
      "At variant site 16869\n",
      "At variant site 16878\n",
      "At variant site 16880\n",
      "At variant site 16926\n",
      "At variant site 16951\n",
      "At variant site 16995\n",
      "At variant site 17001\n",
      "At variant site 17005\n",
      "At variant site 17010\n",
      "At variant site 17030\n",
      "At variant site 17031\n",
      "At variant site 17070\n",
      "At variant site 17065\n",
      "At variant site 17095\n",
      "At variant site 17115\n",
      "At variant site 17137\n",
      "At variant site 17227\n",
      "At variant site 17130\n",
      "At variant site 17250\n",
      "At variant site 17311\n",
      "At variant site 17325\n",
      "At variant site 17401\n",
      "At variant site 17418\n",
      "At variant site 17436\n",
      "At variant site 17322\n",
      "At variant site 17461\n",
      "At variant site 17514\n",
      "At variant site 17577\n",
      "At variant site 17604\n",
      "At variant site 17330\n",
      "At variant site 17665\n",
      "At variant site 17667\n",
      "At variant site 17482\n",
      "At variant site 17669\n",
      "At variant site 17694\n",
      "At variant site 17736\n",
      "At variant site 17737\n",
      "At variant site 17730\n",
      "At variant site 17738\n",
      "At variant site 17757\n",
      "At variant site 17841\n",
      "At variant site 17849\n",
      "At variant site 17850\n",
      "At variant site 17940\n",
      "At variant site 17952\n",
      "At variant site 17976\n",
      "At variant site 18051\n",
      "At variant site 18077\n",
      "At variant site 18096\n",
      "At variant site 18153\n",
      "At variant site 18154\n",
      "At variant site 18162\n",
      "At variant site 18167\n",
      "At variant site 18183\n",
      "At variant site 18243\n",
      "At variant site 18246\n",
      "At variant site 18250\n",
      "At variant site 18297\n",
      "At variant site 18339\n",
      "At variant site 18351\n",
      "At variant site 18402\n",
      "At variant site 18403\n",
      "At variant site 18415\n",
      "At variant site 18477\n",
      "At variant site 18483\n",
      "At variant site 18492\n",
      "At variant site 18494\n",
      "At variant site 18546\n",
      "At variant site 18561\n",
      "At variant site 18574\n",
      "At variant site 18638\n",
      "At variant site 18678\n",
      "At variant site 18694\n",
      "At variant site 18723\n",
      "At variant site 18727\n",
      "At variant site 18731\n",
      "At variant site 18735\n",
      "At variant site 18747\n",
      "At variant site 18852\n",
      "At variant site 18868\n",
      "At variant site 18922\n",
      "At variant site 18970\n",
      "At variant site 19000\n",
      "At variant site 19026\n",
      "At variant site 19054\n",
      "At variant site 19061\n",
      "At variant site 19070\n",
      "At variant site 19088\n",
      "At variant site 19151\n",
      "At variant site 19169\n",
      "At variant site 19171\n",
      "At variant site 19176\n",
      "At variant site 19211\n",
      "At variant site 19317\n",
      "At variant site 19432\n",
      "At variant site 19450\n",
      "At variant site 19471\n",
      "At variant site 19479\n",
      "At variant site 19503\n",
      "At variant site 19515\n",
      "At variant site 19563\n",
      "At variant site 19567\n",
      "At variant site 19584\n",
      "At variant site 19608\n",
      "At variant site 19636\n",
      "At variant site 19668\n",
      "At variant site 19681\n",
      "At variant site 19830\n",
      "At variant site 19723\n",
      "At variant site 19866\n",
      "At variant site 19875\n",
      "At variant site 19877\n",
      "At variant site 19884\n",
      "At variant site 19935\n",
      "At variant site 19946\n",
      "At variant site 19833\n",
      "At variant site 19965\n",
      "At variant site 19990\n",
      "At variant site 20042\n",
      "At variant site 20046\n",
      "At variant site 20051\n",
      "At variant site 20139\n",
      "At variant site 20154\n",
      "At variant site 20169\n",
      "At variant site 20196\n",
      "At variant site 20224\n",
      "At variant site 20235\n",
      "At variant site 20259\n",
      "At variant site 20253\n",
      "At variant site 20266\n",
      "At variant site 20349\n",
      "At variant site 20392\n",
      "At variant site 20387\n",
      "At variant site 20415\n",
      "At variant site 20442\n",
      "At variant site 20518\n",
      "At variant site 20401\n",
      "At variant site 20583\n",
      "At variant site 20661\n",
      "At variant site 20670\n",
      "At variant site 20715\n",
      "At variant site 20732\n",
      "At variant site 20750\n",
      "At variant site 20569\n",
      "At variant site 20753\n",
      "At variant site 20797\n",
      "At variant site 20835\n",
      "At variant site 20906\n",
      "At variant site 20922\n",
      "At variant site 20927\n",
      "At variant site 20955\n",
      "At variant site 20959\n",
      "At variant site 21027\n",
      "At variant site 21048\n",
      "At variant site 21128\n",
      "At variant site 21225\n",
      "At variant site 21246\n",
      "At variant site 21295\n",
      "At variant site 21297\n",
      "At variant site 21334\n",
      "At variant site 21363\n",
      "At variant site 21426\n",
      "At variant site 21459\n",
      "At variant site 21504\n",
      "At variant site 21507\n",
      "At variant site 21540\n",
      "At variant site 21543\n",
      "At variant site 21561\n",
      "At variant site 21566\n",
      "At variant site 21577\n",
      "At variant site 21579\n",
      "At variant site 21586\n",
      "At variant site 21588\n",
      "At variant site 21591\n",
      "At variant site 21595\n",
      "At variant site 21599\n",
      "At variant site 21605\n",
      "At variant site 21606\n",
      "At variant site 21609\n",
      "At variant site 21612\n",
      "At variant site 21613\n",
      "At variant site 21614\n",
      "At variant site 21615\n",
      "At variant site 21618\n",
      "At variant site 21622\n",
      "At variant site 21640\n",
      "At variant site 21658\n",
      "At variant site 21664\n",
      "At variant site 21673\n",
      "At variant site 21693\n",
      "At variant site 21694\n",
      "At variant site 21699\n",
      "At variant site 21700\n",
      "At variant site 21703\n",
      "At variant site 21718\n",
      "At variant site 21734\n",
      "At variant site 21744\n",
      "At variant site 21756\n",
      "At variant site 21758\n",
      "At variant site 21768\n",
      "At variant site 21771\n",
      "At variant site 21778\n",
      "At variant site 21783\n",
      "At variant site 21792\n",
      "At variant site 21828\n",
      "At variant site 21837\n",
      "At variant site 21862\n",
      "At variant site 21877\n",
      "At variant site 21879\n",
      "At variant site 21923\n",
      "At variant site 21934\n",
      "At variant site 21956\n",
      "At variant site 21969\n",
      "At variant site 21977\n",
      "At variant site 21979\n",
      "At variant site 21980\n",
      "At variant site 21982\n",
      "At variant site 21995\n",
      "At variant site 21996\n",
      "At variant site 21997\n",
      "At variant site 21999\n",
      "At variant site 22001\n",
      "At variant site 22011\n",
      "At variant site 22012\n",
      "At variant site 22013\n",
      "At variant site 22014\n",
      "At variant site 22015\n",
      "At variant site 22033\n",
      "At variant site 22067\n",
      "At variant site 22080\n",
      "At variant site 22083\n",
      "At variant site 22085\n",
      "At variant site 22087\n",
      "At variant site 22088\n",
      "At variant site 22094\n",
      "At variant site 22097\n",
      "At variant site 22111\n",
      "At variant site 22129\n",
      "At variant site 22130\n",
      "At variant site 22138\n",
      "At variant site 22141\n",
      "At variant site 22167\n",
      "At variant site 22169\n",
      "At variant site 22176\n",
      "At variant site 22179\n",
      "At variant site 22184\n",
      "At variant site 22185\n",
      "At variant site 22187\n",
      "At variant site 22206\n",
      "At variant site 22308\n",
      "At variant site 22310\n",
      "At variant site 22311\n",
      "At variant site 22312\n",
      "At variant site 22325\n",
      "At variant site 22327\n",
      "At variant site 22332\n",
      "At variant site 22335\n",
      "At variant site 22356\n",
      "At variant site 22406\n",
      "At variant site 22411\n",
      "At variant site 22423\n",
      "At variant site 22429\n",
      "At variant site 22447\n",
      "At variant site 22453\n",
      "At variant site 22459\n",
      "At variant site 22466\n",
      "At variant site 22477\n",
      "At variant site 22529\n",
      "At variant site 22535\n",
      "At variant site 22481\n",
      "At variant site 22557\n",
      "At variant site 22571\n",
      "At variant site 22576\n",
      "At variant site 22549\n",
      "At variant site 22556\n",
      "At variant site 22603\n",
      "At variant site 22578\n",
      "At variant site 22607\n",
      "At variant site 22608\n",
      "At variant site 22643\n",
      "At variant site 22652\n",
      "At variant site 22653\n",
      "At variant site 22658\n",
      "At variant site 22665\n",
      "At variant site 22667\n",
      "At variant site 22668\n",
      "At variant site 22732\n",
      "At variant site 22749\n",
      "At variant site 22754\n",
      "At variant site 22765\n",
      "At variant site 22771\n",
      "At variant site 22777\n",
      "At variant site 22780\n",
      "At variant site 22791\n",
      "At variant site 22792\n",
      "At variant site 22799\n",
      "At variant site 22848\n",
      "At variant site 22861\n",
      "At variant site 22864\n",
      "At variant site 22870\n",
      "At variant site 22872\n",
      "At variant site 22874\n",
      "At variant site 22875\n",
      "At variant site 22877\n",
      "At variant site 22878\n",
      "At variant site 22889\n",
      "At variant site 22895\n",
      "At variant site 22896\n",
      "At variant site 22906\n",
      "At variant site 22907\n",
      "At variant site 22909\n",
      "At variant site 22921\n",
      "At variant site 22936\n",
      "At variant site 22960\n",
      "At variant site 22965\n",
      "At variant site 22966\n",
      "At variant site 22971\n",
      "At variant site 22974\n",
      "At variant site 22984\n",
      "At variant site 22989\n",
      "At variant site 22991\n",
      "At variant site 22992\n",
      "At variant site 22997\n",
      "At variant site 22998\n",
      "At variant site 23009\n",
      "At variant site 23010\n",
      "At variant site 23019\n",
      "At variant site 23021\n",
      "At variant site 23027\n",
      "At variant site 23034\n",
      "At variant site 23042\n",
      "At variant site 23054\n",
      "At variant site 23065\n",
      "At variant site 23072\n",
      "At variant site 23102\n",
      "At variant site 23128\n",
      "At variant site 23134\n",
      "At variant site 23170\n",
      "At variant site 23181\n",
      "At variant site 23187\n",
      "At variant site 23201\n",
      "At variant site 23206\n",
      "At variant site 23224\n",
      "At variant site 23250\n",
      "At variant site 23254\n",
      "At variant site 23257\n",
      "At variant site 23259\n",
      "At variant site 23263\n",
      "At variant site 23266\n",
      "At variant site 23285\n",
      "At variant site 23288\n",
      "At variant site 23320\n",
      "At variant site 23380\n",
      "At variant site 23382\n",
      "At variant site 23396\n",
      "At variant site 23402\n",
      "At variant site 23467\n",
      "At variant site 23504\n",
      "At variant site 23506\n",
      "At variant site 23572\n",
      "At variant site 23578\n",
      "At variant site 23583\n",
      "At variant site 23604\n",
      "At variant site 23617\n",
      "At variant site 23643\n",
      "At variant site 23649\n",
      "At variant site 23652\n",
      "At variant site 23658\n",
      "At variant site 23681\n",
      "At variant site 23686\n",
      "At variant site 23688\n",
      "At variant site 23701\n",
      "At variant site 23709\n",
      "At variant site 23710\n",
      "At variant site 23735\n",
      "At variant site 23744\n",
      "At variant site 23833\n",
      "At variant site 23835\n",
      "At variant site 23875\n",
      "At variant site 23888\n",
      "At variant site 23927\n",
      "At variant site 23965\n",
      "At variant site 23970\n",
      "At variant site 24005\n",
      "At variant site 24013\n",
      "At variant site 24017\n",
      "At variant site 24041\n",
      "At variant site 24109\n",
      "At variant site 24113\n",
      "At variant site 24117\n",
      "At variant site 24166\n",
      "At variant site 24187\n",
      "At variant site 24203\n",
      "At variant site 24236\n",
      "At variant site 24283\n",
      "At variant site 24304\n",
      "At variant site 24307\n",
      "At variant site 24327\n",
      "At variant site 24328\n",
      "At variant site 24316\n",
      "At variant site 24357\n",
      "At variant site 24361\n",
      "At variant site 24389\n",
      "At variant site 24403\n",
      "At variant site 24411\n",
      "At variant site 24421\n",
      "At variant site 24352\n",
      "At variant site 24448\n",
      "At variant site 24478\n",
      "At variant site 24482\n",
      "At variant site 24485\n",
      "At variant site 24565\n",
      "At variant site 24599\n",
      "At variant site 24621\n",
      "At variant site 24673\n",
      "At variant site 24727\n",
      "At variant site 24736\n",
      "At variant site 24754\n",
      "At variant site 24788\n",
      "At variant site 24815\n",
      "At variant site 24842\n",
      "At variant site 24882\n",
      "At variant site 24891\n",
      "At variant site 24893\n",
      "At variant site 24915\n",
      "At variant site 24939\n",
      "At variant site 24969\n",
      "At variant site 24970\n",
      "At variant site 24979\n",
      "At variant site 24982\n",
      "At variant site 24985\n",
      "At variant site 24990\n",
      "At variant site 24998\n",
      "At variant site 25024\n",
      "At variant site 25067\n",
      "At variant site 25075\n",
      "At variant site 25095\n",
      "At variant site 25141\n",
      "At variant site 25158\n",
      "At variant site 25181\n",
      "At variant site 25186\n",
      "At variant site 25208\n",
      "At variant site 25255\n",
      "At variant site 25262\n",
      "At variant site 25318\n",
      "At variant site 25329\n",
      "At variant site 25331\n",
      "At variant site 25338\n",
      "At variant site 25392\n",
      "At variant site 25395\n",
      "At variant site 25136\n",
      "At variant site 25412\n",
      "At variant site 25434\n",
      "At variant site 25440\n",
      "At variant site 25442\n",
      "At variant site 25448\n",
      "At variant site 25470\n",
      "At variant site 25475\n",
      "At variant site 25492\n",
      "At variant site 25493\n",
      "At variant site 25496\n",
      "At variant site 25500\n",
      "At variant site 25505\n",
      "At variant site 25510\n",
      "At variant site 25532\n",
      "At variant site 25534\n",
      "At variant site 25517\n",
      "At variant site 25542\n",
      "At variant site 25544\n",
      "At variant site 25546\n",
      "At variant site 25551\n",
      "At variant site 25563\n",
      "At variant site 25564\n",
      "At variant site 25585\n",
      "At variant site 25588\n",
      "At variant site 25593\n",
      "At variant site 25600\n",
      "At variant site 25603\n",
      "At variant site 25605\n",
      "At variant site 25634\n",
      "At variant site 25637\n",
      "At variant site 25646\n",
      "At variant site 25666\n",
      "At variant site 25680\n",
      "At variant site 25681\n",
      "At variant site 25687\n",
      "At variant site 25689\n",
      "At variant site 25699\n",
      "At variant site 25700\n",
      "At variant site 25739\n",
      "At variant site 25761\n",
      "At variant site 25766\n",
      "At variant site 25785\n",
      "At variant site 25789\n",
      "At variant site 25799\n",
      "At variant site 25805\n",
      "At variant site 25829\n",
      "At variant site 25833\n",
      "At variant site 25860\n",
      "At variant site 25883\n",
      "At variant site 25885\n",
      "At variant site 25834\n",
      "At variant site 25926\n",
      "At variant site 25927\n",
      "At variant site 25938\n",
      "At variant site 25947\n",
      "At variant site 25975\n",
      "At variant site 26008\n",
      "At variant site 25886\n",
      "At variant site 26086\n",
      "At variant site 26090\n",
      "At variant site 26039\n",
      "At variant site 26103\n",
      "At variant site 26123\n",
      "At variant site 26128\n",
      "At variant site 26130\n",
      "At variant site 26131\n",
      "At variant site 26088\n",
      "At variant site 26196\n",
      "At variant site 26201\n",
      "At variant site 26206\n",
      "At variant site 26227\n",
      "At variant site 26249\n",
      "At variant site 26254\n",
      "At variant site 26271\n",
      "At variant site 26284\n",
      "At variant site 26288\n",
      "At variant site 26404\n",
      "At variant site 26435\n",
      "At variant site 26471\n",
      "At variant site 26506\n",
      "At variant site 26508\n",
      "At variant site 26509\n",
      "At variant site 26526\n",
      "At variant site 26556\n",
      "At variant site 26565\n",
      "At variant site 26567\n",
      "At variant site 26585\n",
      "At variant site 26589\n",
      "At variant site 26599\n",
      "At variant site 26619\n",
      "At variant site 26624\n",
      "At variant site 26660\n",
      "At variant site 26679\n",
      "At variant site 26663\n",
      "At variant site 26706\n",
      "At variant site 26723\n",
      "At variant site 26687\n",
      "At variant site 26732\n",
      "At variant site 26746\n",
      "At variant site 26714\n",
      "At variant site 26780\n",
      "At variant site 26729\n",
      "At variant site 26782\n",
      "At variant site 26688\n",
      "At variant site 26812\n",
      "At variant site 26828\n",
      "At variant site 26837\n",
      "At variant site 26855\n",
      "At variant site 26864\n",
      "At variant site 26708\n",
      "At variant site 26798\n",
      "At variant site 26870\n",
      "At variant site 26873\n",
      "At variant site 26874\n",
      "At variant site 26886\n",
      "At variant site 26938\n",
      "At variant site 26943\n",
      "At variant site 26984\n",
      "At variant site 27071\n",
      "At variant site 27122\n",
      "At variant site 27140\n",
      "At variant site 27142\n",
      "At variant site 27185\n",
      "At variant site 27193\n",
      "At variant site 27226\n",
      "At variant site 27238\n",
      "At variant site 27270\n",
      "At variant site 27278\n",
      "At variant site 27282\n",
      "At variant site 27324\n",
      "At variant site 27325\n",
      "At variant site 27361\n",
      "At variant site 27038\n",
      "At variant site 27362\n",
      "At variant site 27363\n",
      "At variant site 27368\n",
      "At variant site 27369\n",
      "At variant site 27380\n",
      "At variant site 27406\n",
      "At variant site 27410\n",
      "At variant site 27413\n",
      "At variant site 27455\n",
      "At variant site 27472\n",
      "At variant site 27473\n",
      "At variant site 27483\n",
      "At variant site 27485\n",
      "At variant site 27486\n",
      "At variant site 27492\n",
      "At variant site 27506\n",
      "At variant site 27538\n",
      "At variant site 27554\n",
      "At variant site 27556\n",
      "At variant site 27572\n",
      "At variant site 27578\n",
      "At variant site 27579\n",
      "At variant site 27582\n",
      "At variant site 27583\n",
      "At variant site 27591\n",
      "At variant site 27617\n",
      "At variant site 27636\n",
      "At variant site 27589\n",
      "At variant site 27651\n",
      "At variant site 27653\n",
      "At variant site 27668\n",
      "At variant site 27679\n",
      "At variant site 27686\n",
      "At variant site 27697\n",
      "At variant site 27700\n",
      "At variant site 27718\n",
      "At variant site 27720\n",
      "At variant site 27731\n",
      "At variant site 27744\n",
      "At variant site 27748\n",
      "At variant site 27764\n",
      "At variant site 27767\n",
      "At variant site 27784\n",
      "At variant site 27786\n",
      "At variant site 27832\n",
      "At variant site 27847\n",
      "At variant site 27853\n",
      "At variant site 27856\n",
      "At variant site 27857\n",
      "At variant site 27860\n",
      "At variant site 27868\n",
      "At variant site 27869\n",
      "At variant site 27876\n",
      "At variant site 27877\n",
      "At variant site 27894\n",
      "At variant site 27898\n",
      "At variant site 27899\n",
      "At variant site 27904\n",
      "At variant site 27916\n",
      "At variant site 27923\n",
      "At variant site 27939\n",
      "At variant site 27943\n",
      "At variant site 27951\n",
      "At variant site 27969\n",
      "At variant site 27978\n",
      "At variant site 27980\n",
      "At variant site 27984\n",
      "At variant site 28027\n",
      "At variant site 28033\n",
      "At variant site 28056\n",
      "At variant site 28058\n",
      "At variant site 28066\n",
      "At variant site 28072\n",
      "At variant site 28074\n",
      "At variant site 28090\n",
      "At variant site 28123\n",
      "At variant site 28146\n",
      "At variant site 28170\n",
      "At variant site 28177\n",
      "At variant site 28188\n",
      "At variant site 28232\n",
      "At variant site 28233\n",
      "At variant site 28250\n",
      "At variant site 28251\n",
      "At variant site 28259\n",
      "At variant site 28260\n",
      "At variant site 28261\n",
      "At variant site 28276\n",
      "At variant site 28278\n",
      "At variant site 28279\n",
      "At variant site 28287\n",
      "At variant site 28290\n",
      "At variant site 28291\n",
      "At variant site 28305\n",
      "At variant site 28309\n",
      "At variant site 28333\n",
      "At variant site 28336\n",
      "At variant site 28341\n",
      "At variant site 28348\n",
      "At variant site 28417\n",
      "At variant site 28423\n",
      "At variant site 28431\n",
      "At variant site 28442\n",
      "At variant site 28482\n",
      "At variant site 28511\n",
      "At variant site 28520\n",
      "At variant site 28501\n",
      "At variant site 28598\n",
      "At variant site 28537\n",
      "At variant site 28627\n",
      "At variant site 28644\n",
      "At variant site 28647\n",
      "At variant site 28651\n",
      "At variant site 28652\n",
      "At variant site 28667\n",
      "At variant site 28687\n",
      "At variant site 28690\n",
      "At variant site 28669\n",
      "At variant site 28694\n",
      "At variant site 28702\n",
      "At variant site 28673\n",
      "At variant site 28709\n",
      "At variant site 28821\n",
      "At variant site 28824\n",
      "At variant site 28831\n",
      "At variant site 28839\n",
      "At variant site 28847\n",
      "At variant site 28848\n",
      "At variant site 28852\n",
      "At variant site 28849\n",
      "At variant site 28851\n",
      "At variant site 28853\n",
      "At variant site 28854\n",
      "At variant site 28857\n",
      "At variant site 28863\n",
      "At variant site 28865\n",
      "At variant site 28866\n",
      "At variant site 28879\n",
      "At variant site 28880\n",
      "At variant site 28883\n",
      "At variant site 28886\n",
      "At variant site 28894\n",
      "At variant site 28898\n",
      "At variant site 28902\n",
      "At variant site 28906\n",
      "At variant site 28928\n",
      "At variant site 28938\n",
      "At variant site 28945\n",
      "At variant site 28947\n",
      "At variant site 28970\n",
      "At variant site 28996\n",
      "At variant site 29020\n",
      "At variant site 29037\n",
      "At variant site 29088\n",
      "At variant site 29089\n",
      "At variant site 29114\n",
      "At variant site 29117\n",
      "At variant site 29167\n",
      "At variant site 29170\n",
      "At variant site 29223\n",
      "At variant site 29242\n",
      "At variant site 29244\n",
      "At variant site 29281\n",
      "At variant site 29317\n",
      "At variant site 29323\n",
      "At variant site 29328\n",
      "At variant site 29332\n",
      "At variant site 29188\n",
      "At variant site 29340\n",
      "At variant site 29369\n",
      "At variant site 29372\n",
      "At variant site 29356\n",
      "At variant site 29375\n",
      "At variant site 29415\n",
      "At variant site 29392\n",
      "At variant site 29436\n",
      "At variant site 29479\n",
      "At variant site 29480\n",
      "At variant site 29483\n",
      "At variant site 29485\n",
      "At variant site 29510\n",
      "At variant site 29511\n",
      "At variant site 29513\n",
      "At variant site 29516\n",
      "At variant site 29528\n",
      "At variant site 29584\n",
      "At variant site 29595\n",
      "At variant site 29598\n",
      "At variant site 29602\n",
      "At variant site 29605\n",
      "At variant site 29615\n",
      "At variant site 29636\n",
      "At variant site 29640\n",
      "At variant site 29649\n",
      "At variant site 29656\n",
      "At variant site 29662\n",
      "At variant site 29666\n",
      "At variant site 29668\n",
      "At variant site 29669\n",
      "At variant site 29670\n",
      "At variant site 29671\n",
      "At variant site 29672\n",
      "At variant site 29673\n",
      "At variant site 29675\n",
      "At variant site 29678\n",
      "At variant site 29679\n",
      "At variant site 29676\n",
      "At variant site 29681\n",
      "At variant site 29685\n",
      "At variant site 29687\n",
      "At variant site 29680\n",
      "At variant site 29694\n",
      "At variant site 29700\n",
      "At variant site 29702\n",
      "At variant site 29704\n",
      "At variant site 29711\n",
      "At variant site 29712\n",
      "At variant site 29713\n",
      "At variant site 29714\n",
      "At variant site 29716\n",
      "At variant site 29717\n",
      "At variant site 29718\n",
      "At variant site 29719\n",
      "At variant site 29720\n",
      "At variant site 29721\n",
      "At variant site 29722\n",
      "At variant site 29723\n",
      "At variant site 29724\n",
      "At variant site 29725\n",
      "At variant site 29726\n",
      "At variant site 29727\n",
      "At variant site 29728\n",
      "At variant site 29729\n",
      "At variant site 29730\n",
      "At variant site 29731\n",
      "At variant site 29732\n",
      "At variant site 29734\n",
      "At variant site 29737\n",
      "At variant site 29738\n",
      "At variant site 29740\n",
      "At variant site 29741\n",
      "At variant site 29742\n",
      "At variant site 29747\n",
      "At variant site 29760\n",
      "At variant site 29761\n",
      "Completed in 76 msec \n",
      "\n",
      "Output newick files will have branch lengths equal to the number of mutations of that branch.\n",
      "\n",
      "Found 0 missing samples.\n",
      "\n",
      "Writing final tree to file /mnt/d/Documents/Masters_Notes/bio_med/code/nosi-phylogenetic-tree/data/cov/final-tree.nh \n",
      "The parsimony score for this tree is: 2182 \n",
      "Completed in 2 msec \n",
      "\n",
      "Saving mutation-annotated tree object to file (after condensing identical sequences) seq_output.nwk\n",
      "Completed in 5 msec \n",
      "\n"
     ]
    }
   ],
   "source": [
    "!usher -t reference_sequences_subset.nwk -v seq_subset.vcf -o seq_output_subset.nwk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ae82cfc-59b3-4ebb-bad4-f656cae6dbf3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_python3",
   "language": "python",
   "name": "conda_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
