{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ce2ca986-b0d9-43b1-8928-00dfbf050a84",
   "metadata": {},
   "source": [
    "# Submodule2: Collect and prepare sequence data and analysis:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "259677a7",
   "metadata": {},
   "source": [
    "# Learning Objectives:\n",
    "In submodule 2 we will leverage the fundamental concepts from [**Submodule 1**](./SubModule1.ipynb) to do the following:\n",
    "- Demonstrate different ways of extracting dataset\n",
    "- Preparing those data for phylogenetic analysis\n",
    "- Discuss how cloud solutions can be helpful on all these things\n",
    "- Implement Quality Control checks on the dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7805b2a7",
   "metadata": {},
   "source": [
    "----------------------------------------------------------------------------------------------------------------\n",
    "# Training Plan \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58931161",
   "metadata": {},
   "source": [
    "Submodule #1: Understanding the Basics of Phylogenetic\n",
    "\n",
    "<font color=\"green\"> **Submodule #2: Collect and Prepare Sequence Data and Analysis** </font>\n",
    "\n",
    "Submodule #3: Construct Phylogenetic Tree\n",
    "\n",
    " \n",
    "Submodule #4: Analyze Phylogenetic Tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a004ca3b",
   "metadata": {},
   "source": [
    "----------------------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4ef0389-80cf-40e1-ac52-c1e75b28ce0e",
   "metadata": {},
   "source": [
    "### 2.1 Demonstrate Efficient Methods for Sourcing Pathogen Sequences and Preparing Data for Phylogenetic Analysis\n",
    "Efficient methods for sourcing pathogen sequences and preparing data include:\n",
    "* Public Databases: Utilize public repositories like GenBank, EMBL, and DDBJ for sourcing high-quality pathogen sequences. These databases provide comprehensive and well-annotated genetic data.\n",
    "* Sequence Retrieval Tools: Use tools like Entrez Direct and Biopython to automate the retrieval of sequences from public databases, ensuring efficiency and reducing the risk of manual errors.\n",
    "* Data Cleaning and Preprocessing: Implement tools such as Trimmomatic for trimming low-quality reads and removing adapters from raw sequence data. This step is crucial for ensuring the integrity of the sequences used in downstream analysis.\n",
    "* Sequence Alignment: Use alignment tools like MAFFT or ClustalW to align sequences, correcting for gaps and mismatches to maximize homology. Proper alignment is essential for accurate phylogenetic tree construction.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbe7bffb-431b-4518-8c64-7ca50ea1e996",
   "metadata": {},
   "source": [
    "### 2.2 Discuss the Importance of Cloud-Based Storage Solutions in Managing Metagenomic Sequence Data\n",
    "Cloud-based storage solutions are vital for managing metagenomic sequence data due to:\n",
    "- **Scalability:** Cloud storage offers scalable solutions to handle the vast amounts of data generated by metagenomic studies. It allows researchers to store large datasets without worrying about local storage limitations.\n",
    "- **Accessibility:** Cloud storage ensures that data can be accessed from anywhere, facilitating collaboration among researchers across different geographical locations.\n",
    "- **Cost-Effectiveness:** Cloud services often operate on a pay-as-you-go model, making it cost-effective as researchers only pay for the storage and computational resources they use.\n",
    "- **Data Security and Backup:** Cloud providers offer robust security measures and automatic backups, protecting valuable data from loss or unauthorized access.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c493fef6-c6b5-4058-8aea-1a094d8feb1f",
   "metadata": {},
   "source": [
    "### 2.3 Explain How the Incorporation of Publicly Available Datasets from Reputable Metagenomic Databases Enhances the Depth of Analysis\n",
    "Incorporating publicly available datasets from reputable metagenomic databases enhances analysis by:\n",
    "- **Increased Data Volume:** Access to a large volume of data increases the statistical power and robustness of analyses, enabling more accurate and comprehensive studies.\n",
    "- **Comparative Analysis:** Public datasets provide a wealth of comparative data, allowing researchers to identify patterns, variations, and evolutionary trends across different studies and datasets.\n",
    "- **Validation and Reproducibility:** Using standardized, publicly available data ensures that results can be validated and reproduced by other researchers, enhancing the credibility of the findings.\n",
    "- **Resource Sharing:** Public databases foster a collaborative environment where researchers can share resources, tools, and data, accelerating the pace of scientific discovery."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f104290f-4fc9-4e61-9473-968c03be0f3d",
   "metadata": {},
   "source": [
    "# Create data directory to store data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8fa8e9f8-1279-4bbf-b4a4-6535e34f5737",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "data_dir = os.path.isdir('./data')\n",
    "if not data_dir:\n",
    "    !mkdir data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbd3190a-88b9-4d99-8b84-ea7f6c45e2bf",
   "metadata": {},
   "source": [
    "### SARS-CoV-2 Dataset\n",
    "Downloaded SARS-CoV-2 sequence data from the Nextstrain project: \n",
    "1. https://docs.nextstrain.org/projects/nextclade/en/stable/user/datasets.html\n",
    "2. The specific file used is: data/nextstrain/sars-cov-2/wuhan-hu-1/proteins/sequences.fasta\n",
    "3. Official Download:\n",
    "    1. Open: https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide&VirusLineage_ss=taxid:2697049&USAState_s=SD\n",
    "    2. Choose any sequence you want and download it as fasta file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "550bbceb-0775-45be-9f13-7a894e40965b",
   "metadata": {},
   "source": [
    "## Download SARS-Cov-2 Dataset from URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "150d58da-941e-4112-8020-4a3fcbc5e6a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory created successfully\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Check if the directory exists\n",
    "cov_dir = os.path.isdir('./data/cov')\n",
    "\n",
    "# If the directory does not exist, create it\n",
    "if not cov_dir:\n",
    "    try:\n",
    "        os.makedirs('./data/cov/')\n",
    "        print(\"Directory created successfully\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2681a615-7aa6-4d9b-995b-89ad842f4c28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/ec2-user/SageMaker/nosi-phylogenetic-tree'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4fa87886-458d-433b-8f24-f715f0995fa4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-09-13 00:18:14--  https://firebasestorage.googleapis.com/v0/b/reactfirebase-142f5.appspot.com/o/nosi-phylogeny%2Fsequences.fasta?alt=media&token=4de7706b-0084-4846-b829-79d8ea35d327\n",
      "Resolving firebasestorage.googleapis.com (firebasestorage.googleapis.com)... 172.253.122.95, 172.253.63.95, 142.250.31.95, ...\n",
      "Connecting to firebasestorage.googleapis.com (firebasestorage.googleapis.com)|172.253.122.95|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4961117 (4.7M) [application/octet-stream]\n",
      "Saving to: ‘./data/cov/sequences.fasta’\n",
      "\n",
      "100%[======================================>] 4,961,117   31.3MB/s   in 0.2s   \n",
      "\n",
      "2024-09-13 00:18:14 (31.3 MB/s) - ‘./data/cov/sequences.fasta’ saved [4961117/4961117]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget -O \"./data/cov/sequences.fasta\" \"https://firebasestorage.googleapis.com/v0/b/reactfirebase-142f5.appspot.com/o/nosi-phylogeny%2Fsequences.fasta?alt=media&token=4de7706b-0084-4846-b829-79d8ea35d327\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "35291ba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download just subset of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a573fce1-a43e-4d28-a910-e1a93ecdba44",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !wget -O \"./data/cov/sequence_PQ230960.fasta\" \"https://firebasestorage.googleapis.com/v0/b/reactfirebase-142f5.appspot.com/o/nosi-phylogeny%2Fsequence_sars_cov_PQ230960.fasta?alt=media&token=ba5ba0c3-a12c-4752-a4bc-63f355761984\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b0fb3b4-9e81-4228-a159-a5bf2b1a9179",
   "metadata": {},
   "source": [
    "### Data Preparation\n",
    "**Convert the .fasta file to .fastq format using the following Python code:**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3673dc77-b348-4f50-992d-4d4dbc4405ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting biopython\n",
      "  Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (12 kB)\n",
      "Requirement already satisfied: numpy in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (from biopython) (1.22.4)\n",
      "Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m65.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: biopython\n",
      "Successfully installed biopython-1.84\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4e74d5fb-729c-4494-b2ee-7330a8be297b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "def fasta_to_fastq(fasta_file, fastq_file, quality=40):\n",
    "    with open(fastq_file, \"w\") as output_handle:\n",
    "        for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "            record.letter_annotations[\"phred_quality\"] = [quality] * len(record.seq)\n",
    "            SeqIO.write(record, output_handle, \"fastq\")\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    fasta_file = \"./data/cov/sequences_subset.fasta\"\n",
    "    fastq_file = \"./data/cov/sequences_subset_converted.fastq\"\n",
    "    fasta_to_fastq(fasta_file=fasta_file, fastq_file=fastq_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2a10fb4-2d75-40b5-8ceb-9fc6b116aa5a",
   "metadata": {},
   "source": [
    "### 2.4 Implement Quality Control Checks Using Tools Like MultiQC and FastQC to Ensure Data Integrity\n",
    "**Using FASTQC for Quality Control.**\n",
    "- FastQC is a java application.  In order to run it needs your system to have a suitable Java Runtime Environment (JRE) installed.  Before you try to run FastQC you should therefore ensure that you have a suitable JRE.  There are a number of different JREs available however the ones we have tested are the latest Oracle runtime environments and those from the adoptOpenJDK project (https://adoptopenjdk.net/).  You need to download and install a suitable 64-bit JRE and make sure that the java application is in your path (most installers will take care of this for you).\n",
    "\n",
    "On linux most distributions will have java installed already so you might not need to do anything.  If java isn't installed then you can add it by doing:\n",
    "\n",
    "- Ubuntu / Mint: **sudo apt install default-jre**\n",
    "\n",
    "- CentOS / Redhat: **sudo yum install java-1.8.0-openjdk**\n",
    "\n",
    "You can check whether java is installed by opening the 'cmd' program on windows, or any shell on linux and typing:\n",
    "\n",
    "**java --version**\n",
    "\n",
    "You should see something like:\n",
    "\n",
    "- java -version\n",
    "- openjdk version \"11.0.2\" 2019-01-15\n",
    "- OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.2+9)\n",
    "- OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.2+9, mixed mode)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8ac211c-d14e-4aea-98a5-f1c68cae643a",
   "metadata": {},
   "source": [
    "### Installation of FastQC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5e65e01-d60c-42d0-8f2d-33f34dde4fd5",
   "metadata": {},
   "source": [
    "### Download and Install FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "51249d58-662f-447c-9c99-e7cd74557184",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10445846-8798-4740-bb95-62aefa58c0cc",
   "metadata": {},
   "source": [
    "### Extract the Downloaded File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5c3a5801-2635-45bb-b75a-477b9fda15a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !unzip fastqc_v0.11.9.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6b4453f-5ab0-42ba-a791-22595baf3c85",
   "metadata": {},
   "source": [
    "### Make the FastQC Executable\n",
    "Navigate to the FastQC directory and make the fastqc file executable:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "76565fbb-f59b-47d0-903c-4367675a940d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %cd FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4021f4d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fcca8708-e663-4706-bfac-79eed92a3b31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !chmod +x fastqc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "06e4842f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving notices: ...working... done\n",
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: - \n",
      "The environment is inconsistent, please check the package plan carefully\n",
      "The following packages are causing the inconsistency:\n",
      "\n",
      "  - conda-forge/noarch::autopep8==2.0.4=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::black==24.4.2=py310hff52083_0\n",
      "  - conda-forge/noarch::bleach==6.1.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::plotly==5.23.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::pytest==8.3.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::qtpy==2.4.1=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::sip==6.7.12=py310hc6cd4ac_0\n",
      "  - conda-forge/noarch::flask==3.0.3=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::importlib_metadata==8.2.0=hd8ed1ab_0\n",
      "  - conda-forge/noarch::lazy_loader==0.4=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::pyqt5-sip==12.12.2=py310hc6cd4ac_5\n",
      "  - conda-forge/noarch::pytoolconfig==1.2.5=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::qdarkstyle==3.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::qtawesome==1.3.1=pyh9208f05_0\n",
      "  - conda-forge/noarch::yapf==0.40.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::dask-core==2024.7.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::flask-cors==4.0.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyter_client==8.6.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::rope==1.13.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::distributed==2024.7.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::ipykernel==6.29.5=pyh3099207_0\n",
      "  - conda-forge/noarch::keyring==25.2.1=pyha804496_0\n",
      "  - conda-forge/noarch::python-lsp-server==1.7.4=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyter_console==6.6.3=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::nbclient==0.10.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::pyls-spyder==0.4.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::python-lsp-black==2.0.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::qtconsole-base==5.4.4=pyha770c72_0\n",
      "  - conda-forge/noarch::s3transfer==0.10.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::spyder-kernels==2.4.4=unix_pyh707e725_0\n",
      "  - conda-forge/noarch::nbconvert-core==7.16.4=pyhd8ed1ab_1\n",
      "  - conda-forge/linux-64::pyqt==5.15.9=py310h04931ad_5\n",
      "  - conda-forge/linux-64::astropy==6.1.0=py310h8a78493_0\n",
      "  - conda-forge/noarch::bokeh==3.5.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyter_server==2.14.2=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::matplotlib-base==3.8.4=py310hef631a5_2\n",
      "  - conda-forge/noarch::nbconvert-pandoc==7.16.4=hd8ed1ab_1\n",
      "  - conda-forge/linux-64::pyqtwebengine==5.15.9=py310h704022c_5\n",
      "  - conda-forge/linux-64::pytables==3.9.2=py310h73b55d5_3\n",
      "  - conda-forge/noarch::qtconsole==5.4.4=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::statsmodels==0.14.2=py310h261611a_0\n",
      "  - conda-forge/noarch::jupyter-lsp==2.2.5=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyterlab_server==2.27.3=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::matplotlib==3.8.4=py310hff52083_2\n",
      "  - conda-forge/noarch::nbconvert==7.16.4=hd8ed1ab_1\n",
      "  - conda-forge/noarch::notebook-shim==0.2.4=pyhd8ed1ab_0\n",
      "  - conda-forge/linux-64::scikit-image==0.24.0=py310hf9f9076_1\n",
      "  - conda-forge/noarch::seaborn-base==0.13.2=pyhd8ed1ab_2\n",
      "  - conda-forge/linux-64::shap==0.45.1=cpu_py310hdd52748_0\n",
      "  - conda-forge/noarch::jupyterlab==4.2.4=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::seaborn==0.13.2=hd8ed1ab_2\n",
      "  - conda-forge/noarch::dask-expr==1.1.9=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::notebook==7.2.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::dask==2024.7.1=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::jupyter==1.0.0=pyhd8ed1ab_10\n",
      "  - conda-forge/noarch::hdijupyterutils==0.21.0=pyh1a96a4e_0\n",
      "  - conda-forge/noarch::autovizwidget==0.21.0=pyh1a96a4e_0\n",
      "  - conda-forge/noarch::sparkmagic==0.21.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::numpydoc==1.7.0=pyhd8ed1ab_3\n",
      "  - conda-forge/linux-64::spyder==5.4.5=py310hff52083_0\n",
      "  - conda-forge/noarch::sphinxcontrib-applehelp==2.0.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinxcontrib-devhelp==2.0.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinxcontrib-htmlhelp==2.1.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinxcontrib-qthelp==2.0.0=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinx==8.0.2=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinxcontrib-serializinghtml==1.1.10=pyhd8ed1ab_0\n",
      "  - conda-forge/noarch::sphinxcontrib-websupport==1.2.7=pyhd8ed1ab_0\n",
      "done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 23.3.1\n",
      "  latest version: 24.7.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "Or to minimize the number of packages updated during conda update use\n",
      "\n",
      "     conda install conda=24.7.1\n",
      "\n",
      "\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/ec2-user/anaconda3/envs/python3\n",
      "\n",
      "  added / updated specs:\n",
      "    - bioconda::fastqc\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    botocore-1.35.23           |pyge310_1234567_0         6.8 MB  conda-forge\n",
      "    fastqc-0.12.1              |       hdfd78af_0        11.1 MB  bioconda\n",
      "    importlib-metadata-8.5.0   |     pyha770c72_0          28 KB  conda-forge\n",
      "    importlib_metadata-8.5.0   |       hd8ed1ab_0           9 KB  conda-forge\n",
      "    openjdk-22.0.1             |       h8651b0f_1       173.3 MB  conda-forge\n",
      "    perl-5.32.1                | 7_hd590300_perl5        12.7 MB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:       204.0 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  botocore           conda-forge/noarch::botocore-1.35.23-pyge310_1234567_0 \n",
      "  cloudpickle        conda-forge/noarch::cloudpickle-3.0.0-pyhd8ed1ab_0 \n",
      "  docutils           conda-forge/noarch::docutils-0.21.2-pyhd8ed1ab_0 \n",
      "  fastqc             bioconda/noarch::fastqc-0.12.1-hdfd78af_0 \n",
      "  importlib-metadata conda-forge/noarch::importlib-metadata-8.5.0-pyha770c72_0 \n",
      "  mock               conda-forge/noarch::mock-5.1.0-pyhd8ed1ab_0 \n",
      "  openjdk            conda-forge/linux-64::openjdk-22.0.1-h8651b0f_1 \n",
      "  packaging          conda-forge/noarch::packaging-24.1-pyhd8ed1ab_0 \n",
      "  perl               conda-forge/linux-64::perl-5.32.1-7_hd590300_perl5 \n",
      "  xorg-libxt         conda-forge/linux-64::xorg-libxt-1.3.0-hd590300_1 \n",
      "\n",
      "The following packages will be UPDATED:\n",
      "\n",
      "  ca-certificates                       2024.7.4-hbcca054_0 --> 2024.8.30-hbcca054_0 \n",
      "  certifi                             2024.7.4-pyhd8ed1ab_0 --> 2024.8.30-pyhd8ed1ab_0 \n",
      "  importlib_metadata                       8.2.0-hd8ed1ab_0 --> 8.5.0-hd8ed1ab_0 \n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "fastqc-0.12.1        | 11.1 MB   |                                       |   0% \n",
      "importlib-metadata-8 | 28 KB     |                                       |   0% \u001b[A\n",
      "\n",
      "importlib_metadata-8 | 9 KB      |                                       |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  |                                       |   0% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "botocore-1.35.23     | 6.8 MB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-5.32.1          | 12.7 MB   |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "importlib-metadata-8 | 28 KB     | #####################1                |  57% \u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  |                                       |   0% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "fastqc-0.12.1        | 11.1 MB   |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-5.32.1          | 12.7 MB   |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | 4                                     |   1% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "fastqc-0.12.1        | 11.1 MB   | ##########1                           |  28% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-5.32.1          | 12.7 MB   | ######8                               |  19% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "botocore-1.35.23     | 6.8 MB    | ##################                    |  49% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "importlib_metadata-8 | 9 KB      | ##################################### | 100% \u001b[A\u001b[A\n",
      "\n",
      "importlib_metadata-8 | 9 KB      | ##################################### | 100% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "fastqc-0.12.1        | 11.1 MB   | ################1                     |  44% \u001b[A\u001b[A\u001b[A\n",
      "importlib-metadata-8 | 28 KB     | ##################################### | 100% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-5.32.1          | 12.7 MB   | ##########9                           |  29% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "botocore-1.35.23     | 6.8 MB    | ########################1             |  65% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "fastqc-0.12.1        | 11.1 MB   | ####################7                 |  56% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-5.32.1          | 12.7 MB   | ################1                     |  44% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "fastqc-0.12.1        | 11.1 MB   | #########################3            |  68% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #1                                    |   3% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-5.32.1          | 12.7 MB   | ####################                  |  54% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "botocore-1.35.23     | 6.8 MB    | ####################################5 |  99% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "fastqc-0.12.1        | 11.1 MB   | #############################6        |  80% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-5.32.1          | 12.7 MB   | #######################7              |  64% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "fastqc-0.12.1        | 11.1 MB   | ###################################7  |  96% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-5.32.1          | 12.7 MB   | ############################3         |  77% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ##2                                   |   6% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-5.32.1          | 12.7 MB   | ###################################4  |  96% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ##6                                   |   7% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###                                   |   8% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###4                                  |   9% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###7                                  |  10% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ####1                                 |  11% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ####5                                 |  12% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #####                                 |  14% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #####5                                |  15% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ######                                |  16% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ######7                               |  18% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #######2                              |  20% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #######7                              |  21% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ########2                             |  22% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ########7                             |  24% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #########1                            |  25% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #########7                            |  26% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ##########1                           |  28% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ##########8                           |  29% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###########3                          |  31% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###########8                          |  32% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ############3                         |  33% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ############8                         |  35% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #############3                        |  36% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #############8                        |  38% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ##############3                       |  39% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ##############7                       |  40% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###############1                      |  41% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###############7                      |  42% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ################1                     |  44% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ################5                     |  45% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ################9                     |  46% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #################3                    |  47% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #################7                    |  48% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ##################2                   |  49% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ##################7                   |  51% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###################2                  |  52% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###################5                  |  53% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###################9                  |  54% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ####################2                 |  55% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ####################6                 |  56% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ####################8                 |  56% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #####################1                |  57% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #####################4                |  58% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "fastqc-0.12.1        | 11.1 MB   | ##################################### | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ######################3               |  60% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ######################7               |  61% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #######################               |  62% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #######################6              |  64% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "botocore-1.35.23     | 6.8 MB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ########################5             |  66% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ##########################1           |  71% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-5.32.1          | 12.7 MB   | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###########################9          |  76% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | #############################6        |  80% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ###############################9      |  86% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ##################################    |  92% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "openjdk-22.0.1       | 173.3 MB  | ####################################5 |  99% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\n",
      "                                                                                \u001b[A\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n"
     ]
    }
   ],
   "source": [
    "!conda install bioconda::fastqc -y"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd4c220f-851a-4f6b-8bfe-b6fc8cf7a70e",
   "metadata": {},
   "source": [
    "### Verify the Installation\n",
    "To verify that FastQC is installed correctly, you can check its version:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "970e4d58-1d66-4308-81d0-a2e213e23f30",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC v0.12.1\r\n"
     ]
    }
   ],
   "source": [
    "!fastqc --version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66cdc713-0d63-48ba-8dde-ed9448bc1306",
   "metadata": {},
   "source": [
    "### Using FastQC\n",
    "You can now use FastQC to analyze your sequence data. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "dea022ee-cab1-4d8b-82ae-2b1682551362",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null\n",
      "Started analysis of sequences_subset_converted.fastq\n",
      "Analysis complete for sequences_subset_converted.fastq\n"
     ]
    }
   ],
   "source": [
    "!fastqc -t 4 ./data/cov/sequences_subset_converted.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ea98ed9-dba4-430c-bd41-d77aaac8f608",
   "metadata": {},
   "source": [
    "# More Datasets to explore"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c551e82f-d6c0-485a-84d8-ba6ab6f04058",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### KEGG Dataset:\n",
    "Dataset 1: KEGG for Phylogenetic Tree\n",
    "Downloading KEGG Dataset\n",
    "**KEGG (Kyoto Encyclopedia of Genes and Genomes)** provides a wealth of data for understanding high-level functions and utilities of biological systems. To download KEGG data:\n",
    "1. Access KEGG Dataset: KEGG provides a website for data retrieval.\n",
    "    - https://www.genome.jp/kegg/seq/\n",
    "2. Downloaded file from FASTA sequence files section on the website.\n",
    "    - More resources on KEGG: https://www.genome.jp/kegg/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a27dde2-52a8-44b4-8463-e5c8412362c8",
   "metadata": {},
   "source": [
    "## Download KEGG dataset from URL"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5f79c10-5ea1-4d86-9648-b5a719572667",
   "metadata": {},
   "source": [
    "### Create Kegg folder to store data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bea11637-3a5f-4ce6-a874-f38f681d2ae0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Check if the directory exists\n",
    "kegg_dir = os.path.isdir('./data/kegg')\n",
    "\n",
    "# If the directory does not exist, create it\n",
    "if not kegg_dir:\n",
    "    try:\n",
    "        os.makedirs('./data/kegg/')\n",
    "        print(\"Directory created successfully\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9ec32755-9a73-48bb-afbb-77cd40cb56f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -O \"./data/kegg/br01553.fasta\" \"https://firebasestorage.googleapis.com/v0/b/reactfirebase-142f5.appspot.com/o/nosi-phylogeny%2Fbr01553.fasta?alt=media&token=08d87736-6f5b-4512-94c8-16531c87aa38\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7b3d154-518a-41c3-afb2-92e8913765e8",
   "metadata": {},
   "source": [
    "## Uniport Dataset\n",
    "Dataset 2: UniProt for Phylogenetic Tree\n",
    "Downloading UniProt Dataset\n",
    "UniProt is a comprehensive resource for protein sequence and functional information. To download UniProt data:\n",
    "1. Access UniProt Website: Visit the UniProt website and search for the desired protein sequences.\n",
    "2. Retrieve Data: Downloaded the Isoform sequences fasta file from the website.\n",
    "    - File name: uniprot_sprot_varsplic.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8de6f7bf-f67e-40c8-8fc1-f56076ad0bf7",
   "metadata": {},
   "source": [
    "## Download Uniport Dataset from URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "22071134-8b6f-4dcd-8d34-2703cc064699",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "Directory created successfully\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Check if the directory exists\n",
    "uniport_dir = os.path.isdir('./data/uniport')\n",
    "\n",
    "# If the directory does not exist, create it\n",
    "if not uniport_dir:\n",
    "    try:\n",
    "        os.makedirs('./data/uniport/')\n",
    "        print(\"Directory created successfully\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c405ffa8-cbfb-416a-afbb-4d12781e01cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget -O \"./data/uniport/uniprot_sprot_varsplic.fasta\" \"https://firebasestorage.googleapis.com/v0/b/reactfirebase-142f5.appspot.com/o/uniprot_sprot_varsplic.fasta?alt=media&token=a7695fee-82a3-4100-983c-5eae5db474e9\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_python3",
   "language": "python",
   "name": "conda_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
